{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bf52b2e2",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "# Assembling\n",
    "\n",
    "This notebook is used to assemble a processable h5ad object for subsequent notebooks."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9e8b8fd",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## Fill in input data, output and settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e99acf34",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Path related settings (these should be the same as for the previous notebook)\n",
    "output_dir = '/mnt/workspace/jdetlef/processed_data' # specify the output directory of the processing should be the same in every notebook\n",
    "test = 'Esophagus_146_0.01' # sample name, this should be the same for all the notebooks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "120c0f22",
   "metadata": {},
   "outputs": [],
   "source": [
    "################### DEFINING STRATEGY ###############################\n",
    "from_h5ad = True #option 1\n",
    "assembling_10_velocity = False #option 2\n",
    "assembling_10_public = False #option 3\n",
    "convert_seurat = False #option 4\n",
    "\n",
    "###################### INPUT DATA ###################################\n",
    "\n",
    "# For option 1: The path to an existing .h5ad file\n",
    "h5ad_files = ['/mnt/workspace/jdetlef/data/anndata/esophagus_muscularis_146_0.01.h5ad']\n",
    "#h5ad_files = ['/mnt/agnerds/PROJECTS/extern/ext442_scATAC_Glaser_11_22/preprocessing_output/data/all_annotated_peaks.h5ad']\n",
    "merge_column='Sample_new' # (str) if multiple h5ad files merged this is an identifier from which file it originates\n",
    "coordinate_cols=None # (list:str) columns where peak location data is stored (['chr', 'start', 'end'])\n",
    "set_index=True # (boolean) should the adata.var index be formatted, that it matches chr:start-stop\n",
    "index_from=\"name\" # (str) should the index be generated from a certain column. Otherwise this is None (SnapATAC: name)\n",
    "\n",
    "# Manually set existing QC Columns\n",
    "n_features_by_counts = None\n",
    "log1p_n_features_by_counts = None\n",
    "total_counts = None\n",
    "log1p_total_counts = None\n",
    "mean_insertsize = None\n",
    "insertsize_count = None\n",
    "n_total_fragments = None\n",
    "n_fragments_in_promoters = None\n",
    "pct_fragments_in_promoters = None\n",
    "blacklist_overlaps = None\n",
    "# total_number_of_fragments\n",
    "TN = None # 'TN'\n",
    "# uniquely_mapped_fragments\n",
    "UM = None # 'UM'\n",
    "# properly_paired_fragments\n",
    "PP = None # 'PP'\n",
    "# uniq_fragments\n",
    "UQ = None # 'UQ'\n",
    "# chrM_fragments\n",
    "CM = None # 'CM'\n",
    "\n",
    "#################### Calc QC Columns ###############################\n",
    "# set available number of threads\n",
    "n_threads = 8\n",
    "\n",
    "# Step 1: QC Metrics\n",
    "calc_qc_metrics = True # set True if the QC metrics should be calculated \n",
    "\n",
    "# Step 2 & 3\n",
    "use_bam = True # False if fragments should be used\n",
    "barcode_tag = 'CB' # set the tag where to find the barcode to use\n",
    "# if Step 2, 3 or both is executed specify either a BAM file or a fragments file:\n",
    "bam_file =  '/mnt/workspace/jdetlef/data/bamfiles/sorted_esophagus_muscularis_146_0.01.bam' # specify bamfile of the corresponding alignment\n",
    "fragments_file = '/mnt/workspace/jdetlef/data/bamfiles/fragments_esophagus_muscularis_146_0.01.bed' # specify fragments file (if this is possible use the fragments file directly)\n",
    "\n",
    "# Step 2: Mean Insertsizes\n",
    "calc_mean_insertsize = True # set True if the mean insertsize should be calculated\n",
    "\n",
    "# Step 3: Promotor-enrichment\n",
    "calc_promotor_enrichment = True # set True if the promotor-enrichment should be calculated\n",
    "# specify promotors_gtf if another promoter than the fixed ones should be used\n",
    "# if promotors is None please specify species\n",
    "promoters_gtf =  '/mnt/workspace/jdetlef/data/homo_sapiens.104.promoters2000.gtf' # '/mnt/workspace/jdetlef/ext_ana/mus_musculus.104.promoters2000.gtf'\n",
    "species = None"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d6dd817",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## Import modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c997211a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sctoolbox modules \n",
    "import sctoolbox.atac_tree as sub_tree\n",
    "import sctoolbox.creators as cr\n",
    "import sctoolbox.fragment_length as fragments\n",
    "import sctoolbox.atac as atac\n",
    "import sctoolbox.calc_overlap_pct as overlap\n",
    "import sctoolbox.analyser \n",
    "import sctoolbox.nucleosome_utils as nuc_utils\n",
    "from sctoolbox.qc_filter import *\n",
    "from sctoolbox.atac_utils import *\n",
    "# import episcanpy\n",
    "import episcanpy as epi"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fc0e3d7",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## Setup path handling object "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "880c04fb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: run is None\n",
      "all directories existing\n",
      "all directories existing\n"
     ]
    }
   ],
   "source": [
    "# make an instance of the class\n",
    "tree = sub_tree.ATAC_tree()\n",
    "# set processing/output directory\n",
    "tree.processing_dir = output_dir\n",
    "# set sample/experiment.. \n",
    "tree.run = test"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84b9cddf",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## Read in data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e7b43d6",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### Option 1: Read from .h5ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "96a9a071",
   "metadata": {},
   "outputs": [],
   "source": [
    "qc_columns = {}\n",
    "qc_columns[\"n_features_by_counts\"] = n_features_by_counts\n",
    "qc_columns[\"log1p_n_features_by_counts\"] = log1p_n_features_by_counts\n",
    "qc_columns[\"total_counts\"] = total_counts\n",
    "qc_columns[\"log1p_total_counts\"] = log1p_total_counts\n",
    "qc_columns[\"mean_insertsize\"] = mean_insertsize\n",
    "qc_columns['n_total_fragments'] = n_total_fragments\n",
    "qc_columns['n_fragments_in_promoters'] = n_fragments_in_promoters\n",
    "qc_columns['pct_fragments_in_promoters'] = pct_fragments_in_promoters\n",
    "qc_columns[\"blacklist_overlaps\"] = blacklist_overlaps\n",
    "qc_columns[\"TN\"] = TN\n",
    "qc_columns[\"UM\"] = UM\n",
    "qc_columns[\"PP\"] = PP\n",
    "qc_columns[\"UQ\"] = UQ\n",
    "qc_columns[\"CM\"] = CM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6b900535",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " ImplicitModificationWarning:/home/jdetlef/.conda/envs/sctoolbox/lib/python3.10/site-packages/anndata/_core/anndata.py:121: Transforming to str index.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['peak_chr', 'peak_start', 'peak_end'], dtype='object')\n",
      "add existing adata.obs columns to infoprocess:\n",
      "\n",
      "setting adata.obs.index = adata.obs[barcode]\n"
     ]
    }
   ],
   "source": [
    "if from_h5ad:\n",
    "    \n",
    "    adata = assemble_from_h5ad(h5ad_files,\n",
    "                       qc_columns,\n",
    "                       merge_column=merge_column,\n",
    "                       coordinate_cols=coordinate_cols,\n",
    "                       set_index=set_index,\n",
    "                       index_from=index_from)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "746f9da5",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## Inspect adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9e059525",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 73652 × 174262\n",
       "    obs: 'TN', 'UM', 'PP', 'UQ', 'CM', 'file', 'sample', 'Sample_new'\n",
       "    var: 'peak_chr', 'peak_start', 'peak_end'\n",
       "    uns: 'infoprocess', 'color_set'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d4ec8cb1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TN</th>\n",
       "      <th>UM</th>\n",
       "      <th>PP</th>\n",
       "      <th>UQ</th>\n",
       "      <th>CM</th>\n",
       "      <th>file</th>\n",
       "      <th>sample</th>\n",
       "      <th>Sample_new</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACCCGAGATA</th>\n",
       "      <td>33.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACCTAAGTGG</th>\n",
       "      <td>52.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACGGATCAGT</th>\n",
       "      <td>27.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACGTCCCGTT</th>\n",
       "      <td>6223.0</td>\n",
       "      <td>5231.0</td>\n",
       "      <td>5213.0</td>\n",
       "      <td>3779.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACTAGCCCTA</th>\n",
       "      <td>41.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTCGCGTGTAA</th>\n",
       "      <td>9.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTCTGCAGACT</th>\n",
       "      <td>10.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTCTGGCGCAG</th>\n",
       "      <td>17.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTGCCGGAAGG</th>\n",
       "      <td>8.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTGTGTTACCG</th>\n",
       "      <td>19.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>73652 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            TN      UM      PP      UQ    CM  \\\n",
       "AAACTACCAGAAACCCGAGATA    33.0    18.0    18.0    15.0   0.0   \n",
       "AAACTACCAGAAACCTAAGTGG    52.0    36.0    35.0    32.0   1.0   \n",
       "AAACTACCAGAAACGGATCAGT    27.0    19.0    19.0    19.0   0.0   \n",
       "AAACTACCAGAAACGTCCCGTT  6223.0  5231.0  5213.0  3779.0  18.0   \n",
       "AAACTACCAGAAACTAGCCCTA    41.0    29.0    29.0    26.0   0.0   \n",
       "...                        ...     ...     ...     ...   ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA     9.0     6.0     6.0     6.0   0.0   \n",
       "TTCCATCCTCTTTCTGCAGACT    10.0     6.0     6.0     6.0   1.0   \n",
       "TTCCATCCTCTTTCTGGCGCAG    17.0     5.0     5.0     5.0   1.0   \n",
       "TTCCATCCTCTTTGCCGGAAGG     8.0     7.0     7.0     7.0   0.0   \n",
       "TTCCATCCTCTTTGTGTTACCG    19.0     9.0     9.0     7.0   0.0   \n",
       "\n",
       "                                                                     file  \\\n",
       "AAACTACCAGAAACCCGAGATA  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "AAACTACCAGAAACCTAAGTGG  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "AAACTACCAGAAACGGATCAGT  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "AAACTACCAGAAACGTCCCGTT  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "AAACTACCAGAAACTAGCCCTA  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "...                                                                   ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "TTCCATCCTCTTTCTGCAGACT  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "TTCCATCCTCTTTCTGGCGCAG  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "TTCCATCCTCTTTGCCGGAAGG  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "TTCCATCCTCTTTGTGTTACCG  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "\n",
       "                       sample Sample_new  \n",
       "AAACTACCAGAAACCCGAGATA     MB    sample1  \n",
       "AAACTACCAGAAACCTAAGTGG     MB    sample1  \n",
       "AAACTACCAGAAACGGATCAGT     MB    sample1  \n",
       "AAACTACCAGAAACGTCCCGTT     MB    sample1  \n",
       "AAACTACCAGAAACTAGCCCTA     MB    sample1  \n",
       "...                       ...        ...  \n",
       "TTCCATCCTCTTTCGCGTGTAA     MB    sample1  \n",
       "TTCCATCCTCTTTCTGCAGACT     MB    sample1  \n",
       "TTCCATCCTCTTTCTGGCGCAG     MB    sample1  \n",
       "TTCCATCCTCTTTGCCGGAAGG     MB    sample1  \n",
       "TTCCATCCTCTTTGTGTTACCG     MB    sample1  \n",
       "\n",
       "[73652 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(adata.obs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3405de5d",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## Calculate missing columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78fcd9b3",
   "metadata": {},
   "source": [
    "### 1. calculate scanpy qc_metrics (n_features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "97a9760e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TN</th>\n",
       "      <th>UM</th>\n",
       "      <th>PP</th>\n",
       "      <th>UQ</th>\n",
       "      <th>CM</th>\n",
       "      <th>file</th>\n",
       "      <th>sample</th>\n",
       "      <th>Sample_new</th>\n",
       "      <th>n_features_by_counts</th>\n",
       "      <th>log1p_n_features_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACCCGAGATA</th>\n",
       "      <td>33.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>5</td>\n",
       "      <td>1.791759</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.791759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACCTAAGTGG</th>\n",
       "      <td>52.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>17</td>\n",
       "      <td>2.890372</td>\n",
       "      <td>17.0</td>\n",
       "      <td>2.890372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACGGATCAGT</th>\n",
       "      <td>27.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>10</td>\n",
       "      <td>2.397895</td>\n",
       "      <td>10.0</td>\n",
       "      <td>2.397895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACGTCCCGTT</th>\n",
       "      <td>6223.0</td>\n",
       "      <td>5231.0</td>\n",
       "      <td>5213.0</td>\n",
       "      <td>3779.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>1648</td>\n",
       "      <td>7.407924</td>\n",
       "      <td>1791.0</td>\n",
       "      <td>7.491087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACTAGCCCTA</th>\n",
       "      <td>41.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>13</td>\n",
       "      <td>2.639057</td>\n",
       "      <td>13.0</td>\n",
       "      <td>2.639057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTCGCGTGTAA</th>\n",
       "      <td>9.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>3</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.386294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTCTGCAGACT</th>\n",
       "      <td>10.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>3</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.386294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTCTGGCGCAG</th>\n",
       "      <td>17.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTGCCGGAAGG</th>\n",
       "      <td>8.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>5</td>\n",
       "      <td>1.791759</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.791759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTGTGTTACCG</th>\n",
       "      <td>19.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>2</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.098612</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>73652 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            TN      UM      PP      UQ    CM  \\\n",
       "AAACTACCAGAAACCCGAGATA    33.0    18.0    18.0    15.0   0.0   \n",
       "AAACTACCAGAAACCTAAGTGG    52.0    36.0    35.0    32.0   1.0   \n",
       "AAACTACCAGAAACGGATCAGT    27.0    19.0    19.0    19.0   0.0   \n",
       "AAACTACCAGAAACGTCCCGTT  6223.0  5231.0  5213.0  3779.0  18.0   \n",
       "AAACTACCAGAAACTAGCCCTA    41.0    29.0    29.0    26.0   0.0   \n",
       "...                        ...     ...     ...     ...   ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA     9.0     6.0     6.0     6.0   0.0   \n",
       "TTCCATCCTCTTTCTGCAGACT    10.0     6.0     6.0     6.0   1.0   \n",
       "TTCCATCCTCTTTCTGGCGCAG    17.0     5.0     5.0     5.0   1.0   \n",
       "TTCCATCCTCTTTGCCGGAAGG     8.0     7.0     7.0     7.0   0.0   \n",
       "TTCCATCCTCTTTGTGTTACCG    19.0     9.0     9.0     7.0   0.0   \n",
       "\n",
       "                                                                     file  \\\n",
       "AAACTACCAGAAACCCGAGATA  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "AAACTACCAGAAACCTAAGTGG  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "AAACTACCAGAAACGGATCAGT  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "AAACTACCAGAAACGTCCCGTT  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "AAACTACCAGAAACTAGCCCTA  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "...                                                                   ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "TTCCATCCTCTTTCTGCAGACT  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "TTCCATCCTCTTTCTGGCGCAG  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "TTCCATCCTCTTTGCCGGAAGG  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "TTCCATCCTCTTTGTGTTACCG  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "\n",
       "                       sample Sample_new  n_features_by_counts  \\\n",
       "AAACTACCAGAAACCCGAGATA     MB    sample1                     5   \n",
       "AAACTACCAGAAACCTAAGTGG     MB    sample1                    17   \n",
       "AAACTACCAGAAACGGATCAGT     MB    sample1                    10   \n",
       "AAACTACCAGAAACGTCCCGTT     MB    sample1                  1648   \n",
       "AAACTACCAGAAACTAGCCCTA     MB    sample1                    13   \n",
       "...                       ...        ...                   ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA     MB    sample1                     3   \n",
       "TTCCATCCTCTTTCTGCAGACT     MB    sample1                     3   \n",
       "TTCCATCCTCTTTCTGGCGCAG     MB    sample1                     0   \n",
       "TTCCATCCTCTTTGCCGGAAGG     MB    sample1                     5   \n",
       "TTCCATCCTCTTTGTGTTACCG     MB    sample1                     2   \n",
       "\n",
       "                        log1p_n_features_by_counts  total_counts  \\\n",
       "AAACTACCAGAAACCCGAGATA                    1.791759           5.0   \n",
       "AAACTACCAGAAACCTAAGTGG                    2.890372          17.0   \n",
       "AAACTACCAGAAACGGATCAGT                    2.397895          10.0   \n",
       "AAACTACCAGAAACGTCCCGTT                    7.407924        1791.0   \n",
       "AAACTACCAGAAACTAGCCCTA                    2.639057          13.0   \n",
       "...                                            ...           ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA                    1.386294           3.0   \n",
       "TTCCATCCTCTTTCTGCAGACT                    1.386294           3.0   \n",
       "TTCCATCCTCTTTCTGGCGCAG                    0.000000           0.0   \n",
       "TTCCATCCTCTTTGCCGGAAGG                    1.791759           5.0   \n",
       "TTCCATCCTCTTTGTGTTACCG                    1.098612           2.0   \n",
       "\n",
       "                        log1p_total_counts  \n",
       "AAACTACCAGAAACCCGAGATA            1.791759  \n",
       "AAACTACCAGAAACCTAAGTGG            2.890372  \n",
       "AAACTACCAGAAACGGATCAGT            2.397895  \n",
       "AAACTACCAGAAACGTCCCGTT            7.491087  \n",
       "AAACTACCAGAAACTAGCCCTA            2.639057  \n",
       "...                                    ...  \n",
       "TTCCATCCTCTTTCGCGTGTAA            1.386294  \n",
       "TTCCATCCTCTTTCTGCAGACT            1.386294  \n",
       "TTCCATCCTCTTTCTGGCGCAG            0.000000  \n",
       "TTCCATCCTCTTTGCCGGAAGG            1.791759  \n",
       "TTCCATCCTCTTTGTGTTACCG            1.098612  \n",
       "\n",
       "[73652 rows x 12 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if qc_columns['n_features_by_counts'] is None or qc_columns['log1p_n_features_by_counts'] is None and calc_qc_metrics:\n",
    "    adata = analyser.calculate_qc_metrics(adata, var_type='features')\n",
    "    \n",
    "    qc_columns['n_features_by_counts'] = 'n_features_by_counts'\n",
    "    qc_columns['log1p_n_features_by_counts'] = 'log1p_n_features_by_counts'\n",
    "    qc_columns['total_counts'] = 'total_counts'\n",
    "    qc_columns['log1p_total_counts'] = 'log1p_total_counts'\n",
    "    \n",
    "    build_legend(adata, 'n_features_by_counts', 'n_features_by_counts')\n",
    "    build_legend(adata, 'log1p_n_features_by_counts', 'log1p_n_features_by_counts')\n",
    "    build_legend(adata, 'total_counts', 'total_counts')\n",
    "    build_legend(adata, 'log1p_total_counts', 'log1p_total_counts')\n",
    "\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a26513bb",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### check barcode tag"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "860636d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Barcode tag: OK\n"
     ]
    }
   ],
   "source": [
    "check_barcode_tag(adata, bam_file, barcode_tag)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5d66924",
   "metadata": {},
   "source": [
    "### 2. calc mean insertsize and count if missing "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "114893bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading fragments...\n",
      "Time for loading fragments:  108.36348223686218\n",
      "calculate distribution...\n",
      "Time for calculating distribution:  24.630712747573853\n",
      "calculate maxima...\n",
      "Time for calculating maxima:  7.017499208450317\n",
      "calculate score...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " RuntimeWarning:/home/jdetlef/.conda/envs/sctoolbox/lib/python3.10/site-packages/sctoolbox/nucleosome_utils.py:248: divide by zero encountered in double_scalars\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time for calculating score:  17.567495346069336\n"
     ]
    }
   ],
   "source": [
    "adata = nuc_utils.add_chromatin_conditions(adata, fragments_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "bd633f0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TN</th>\n",
       "      <th>UM</th>\n",
       "      <th>PP</th>\n",
       "      <th>UQ</th>\n",
       "      <th>CM</th>\n",
       "      <th>file</th>\n",
       "      <th>sample</th>\n",
       "      <th>Sample_new</th>\n",
       "      <th>n_features_by_counts</th>\n",
       "      <th>log1p_n_features_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>Mean-Fragment-Length</th>\n",
       "      <th>Median-Fragment-Length</th>\n",
       "      <th>Fragments</th>\n",
       "      <th>Fragment-Count</th>\n",
       "      <th>Distribution</th>\n",
       "      <th>Maxima</th>\n",
       "      <th>Maxima-Count</th>\n",
       "      <th>nucleosomal-score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACCCGAGATA</th>\n",
       "      <td>33.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>5</td>\n",
       "      <td>1.791759</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.791759</td>\n",
       "      <td>82.36</td>\n",
       "      <td>49</td>\n",
       "      <td>[123, 46, 90, 49, 201, 61, 178, 44, 34, 42, 38]</td>\n",
       "      <td>11.0</td>\n",
       "      <td>[9, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>145.544305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACCTAAGTGG</th>\n",
       "      <td>52.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>17</td>\n",
       "      <td>2.890372</td>\n",
       "      <td>17.0</td>\n",
       "      <td>2.890372</td>\n",
       "      <td>170.38</td>\n",
       "      <td>138.5</td>\n",
       "      <td>[72, 417, 148, 40, 36, 129, 207, 89, 166, 111,...</td>\n",
       "      <td>26.0</td>\n",
       "      <td>[16, 7, 2, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...</td>\n",
       "      <td>[0]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>102.513844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACGGATCAGT</th>\n",
       "      <td>27.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>10</td>\n",
       "      <td>2.397895</td>\n",
       "      <td>10.0</td>\n",
       "      <td>2.397895</td>\n",
       "      <td>161.17</td>\n",
       "      <td>74.0</td>\n",
       "      <td>[407, 78, 50, 52, 26, 70, 81, 178, 496, 28, 53...</td>\n",
       "      <td>12.0</td>\n",
       "      <td>[8, 1, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>140.045502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACGTCCCGTT</th>\n",
       "      <td>6223.0</td>\n",
       "      <td>5231.0</td>\n",
       "      <td>5213.0</td>\n",
       "      <td>3779.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>1648</td>\n",
       "      <td>7.407924</td>\n",
       "      <td>1791.0</td>\n",
       "      <td>7.491087</td>\n",
       "      <td>144.03</td>\n",
       "      <td>106</td>\n",
       "      <td>[60, 54, 31, 201, 41, 267, 49, 259, 275, 77, 6...</td>\n",
       "      <td>3475.0</td>\n",
       "      <td>[2351, 871, 229, 20, 4, 0, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>[0]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>19.623836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACTAGCCCTA</th>\n",
       "      <td>41.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>13</td>\n",
       "      <td>2.639057</td>\n",
       "      <td>13.0</td>\n",
       "      <td>2.639057</td>\n",
       "      <td>160.5</td>\n",
       "      <td>131.5</td>\n",
       "      <td>[56, 350, 37, 391, 212, 40, 207, 48, 49, 215]</td>\n",
       "      <td>10.0</td>\n",
       "      <td>[5, 3, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>152.003069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTCGCGTGTAA</th>\n",
       "      <td>9.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>3</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>125.8</td>\n",
       "      <td>94</td>\n",
       "      <td>[91, 94, 199, 77, 168]</td>\n",
       "      <td>5.0</td>\n",
       "      <td>[4, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>220.573902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTCTGCAGACT</th>\n",
       "      <td>10.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>3</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>87.0</td>\n",
       "      <td>58</td>\n",
       "      <td>[58, 147, 56]</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>324.955404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTCTGGCGCAG</th>\n",
       "      <td>17.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>86.0</td>\n",
       "      <td>46</td>\n",
       "      <td>[176, 36, 46]</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[2, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>324.955404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTGCCGGAAGG</th>\n",
       "      <td>8.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>5</td>\n",
       "      <td>1.791759</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.791759</td>\n",
       "      <td>163.5</td>\n",
       "      <td>163.5</td>\n",
       "      <td>[266, 61]</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTGTGTTACCG</th>\n",
       "      <td>19.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>2</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>115.0</td>\n",
       "      <td>100</td>\n",
       "      <td>[122, 46, 241, 100, 66]</td>\n",
       "      <td>5.0</td>\n",
       "      <td>[4, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>220.573902</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>73652 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            TN      UM      PP      UQ    CM  \\\n",
       "AAACTACCAGAAACCCGAGATA    33.0    18.0    18.0    15.0   0.0   \n",
       "AAACTACCAGAAACCTAAGTGG    52.0    36.0    35.0    32.0   1.0   \n",
       "AAACTACCAGAAACGGATCAGT    27.0    19.0    19.0    19.0   0.0   \n",
       "AAACTACCAGAAACGTCCCGTT  6223.0  5231.0  5213.0  3779.0  18.0   \n",
       "AAACTACCAGAAACTAGCCCTA    41.0    29.0    29.0    26.0   0.0   \n",
       "...                        ...     ...     ...     ...   ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA     9.0     6.0     6.0     6.0   0.0   \n",
       "TTCCATCCTCTTTCTGCAGACT    10.0     6.0     6.0     6.0   1.0   \n",
       "TTCCATCCTCTTTCTGGCGCAG    17.0     5.0     5.0     5.0   1.0   \n",
       "TTCCATCCTCTTTGCCGGAAGG     8.0     7.0     7.0     7.0   0.0   \n",
       "TTCCATCCTCTTTGTGTTACCG    19.0     9.0     9.0     7.0   0.0   \n",
       "\n",
       "                                                                     file  \\\n",
       "AAACTACCAGAAACCCGAGATA  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "AAACTACCAGAAACCTAAGTGG  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "AAACTACCAGAAACGGATCAGT  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "AAACTACCAGAAACGTCCCGTT  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "AAACTACCAGAAACTAGCCCTA  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "...                                                                   ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "TTCCATCCTCTTTCTGCAGACT  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "TTCCATCCTCTTTCTGGCGCAG  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "TTCCATCCTCTTTGCCGGAAGG  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "TTCCATCCTCTTTGTGTTACCG  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "\n",
       "                       sample Sample_new  n_features_by_counts  \\\n",
       "AAACTACCAGAAACCCGAGATA     MB    sample1                     5   \n",
       "AAACTACCAGAAACCTAAGTGG     MB    sample1                    17   \n",
       "AAACTACCAGAAACGGATCAGT     MB    sample1                    10   \n",
       "AAACTACCAGAAACGTCCCGTT     MB    sample1                  1648   \n",
       "AAACTACCAGAAACTAGCCCTA     MB    sample1                    13   \n",
       "...                       ...        ...                   ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA     MB    sample1                     3   \n",
       "TTCCATCCTCTTTCTGCAGACT     MB    sample1                     3   \n",
       "TTCCATCCTCTTTCTGGCGCAG     MB    sample1                     0   \n",
       "TTCCATCCTCTTTGCCGGAAGG     MB    sample1                     5   \n",
       "TTCCATCCTCTTTGTGTTACCG     MB    sample1                     2   \n",
       "\n",
       "                        log1p_n_features_by_counts  total_counts  \\\n",
       "AAACTACCAGAAACCCGAGATA                    1.791759           5.0   \n",
       "AAACTACCAGAAACCTAAGTGG                    2.890372          17.0   \n",
       "AAACTACCAGAAACGGATCAGT                    2.397895          10.0   \n",
       "AAACTACCAGAAACGTCCCGTT                    7.407924        1791.0   \n",
       "AAACTACCAGAAACTAGCCCTA                    2.639057          13.0   \n",
       "...                                            ...           ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA                    1.386294           3.0   \n",
       "TTCCATCCTCTTTCTGCAGACT                    1.386294           3.0   \n",
       "TTCCATCCTCTTTCTGGCGCAG                    0.000000           0.0   \n",
       "TTCCATCCTCTTTGCCGGAAGG                    1.791759           5.0   \n",
       "TTCCATCCTCTTTGTGTTACCG                    1.098612           2.0   \n",
       "\n",
       "                        log1p_total_counts Mean-Fragment-Length  \\\n",
       "AAACTACCAGAAACCCGAGATA            1.791759                82.36   \n",
       "AAACTACCAGAAACCTAAGTGG            2.890372               170.38   \n",
       "AAACTACCAGAAACGGATCAGT            2.397895               161.17   \n",
       "AAACTACCAGAAACGTCCCGTT            7.491087               144.03   \n",
       "AAACTACCAGAAACTAGCCCTA            2.639057                160.5   \n",
       "...                                    ...                  ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA            1.386294                125.8   \n",
       "TTCCATCCTCTTTCTGCAGACT            1.386294                 87.0   \n",
       "TTCCATCCTCTTTCTGGCGCAG            0.000000                 86.0   \n",
       "TTCCATCCTCTTTGCCGGAAGG            1.791759                163.5   \n",
       "TTCCATCCTCTTTGTGTTACCG            1.098612                115.0   \n",
       "\n",
       "                       Median-Fragment-Length  \\\n",
       "AAACTACCAGAAACCCGAGATA                     49   \n",
       "AAACTACCAGAAACCTAAGTGG                  138.5   \n",
       "AAACTACCAGAAACGGATCAGT                   74.0   \n",
       "AAACTACCAGAAACGTCCCGTT                    106   \n",
       "AAACTACCAGAAACTAGCCCTA                  131.5   \n",
       "...                                       ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA                     94   \n",
       "TTCCATCCTCTTTCTGCAGACT                     58   \n",
       "TTCCATCCTCTTTCTGGCGCAG                     46   \n",
       "TTCCATCCTCTTTGCCGGAAGG                  163.5   \n",
       "TTCCATCCTCTTTGTGTTACCG                    100   \n",
       "\n",
       "                                                                Fragments  \\\n",
       "AAACTACCAGAAACCCGAGATA    [123, 46, 90, 49, 201, 61, 178, 44, 34, 42, 38]   \n",
       "AAACTACCAGAAACCTAAGTGG  [72, 417, 148, 40, 36, 129, 207, 89, 166, 111,...   \n",
       "AAACTACCAGAAACGGATCAGT  [407, 78, 50, 52, 26, 70, 81, 178, 496, 28, 53...   \n",
       "AAACTACCAGAAACGTCCCGTT  [60, 54, 31, 201, 41, 267, 49, 259, 275, 77, 6...   \n",
       "AAACTACCAGAAACTAGCCCTA      [56, 350, 37, 391, 212, 40, 207, 48, 49, 215]   \n",
       "...                                                                   ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA                             [91, 94, 199, 77, 168]   \n",
       "TTCCATCCTCTTTCTGCAGACT                                      [58, 147, 56]   \n",
       "TTCCATCCTCTTTCTGGCGCAG                                      [176, 36, 46]   \n",
       "TTCCATCCTCTTTGCCGGAAGG                                          [266, 61]   \n",
       "TTCCATCCTCTTTGTGTTACCG                            [122, 46, 241, 100, 66]   \n",
       "\n",
       "                        Fragment-Count  \\\n",
       "AAACTACCAGAAACCCGAGATA            11.0   \n",
       "AAACTACCAGAAACCTAAGTGG            26.0   \n",
       "AAACTACCAGAAACGGATCAGT            12.0   \n",
       "AAACTACCAGAAACGTCCCGTT          3475.0   \n",
       "AAACTACCAGAAACTAGCCCTA            10.0   \n",
       "...                                ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA             5.0   \n",
       "TTCCATCCTCTTTCTGCAGACT             3.0   \n",
       "TTCCATCCTCTTTCTGGCGCAG             3.0   \n",
       "TTCCATCCTCTTTGCCGGAAGG             2.0   \n",
       "TTCCATCCTCTTTGTGTTACCG             5.0   \n",
       "\n",
       "                                                             Distribution  \\\n",
       "AAACTACCAGAAACCCGAGATA  [9, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "AAACTACCAGAAACCTAAGTGG  [16, 7, 2, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...   \n",
       "AAACTACCAGAAACGGATCAGT  [8, 1, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "AAACTACCAGAAACGTCCCGTT  [2351, 871, 229, 20, 4, 0, 0, 0, 0, 0, 0, 0, 0...   \n",
       "AAACTACCAGAAACTAGCCCTA  [5, 3, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "...                                                                   ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA  [4, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "TTCCATCCTCTTTCTGCAGACT  [3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "TTCCATCCTCTTTCTGGCGCAG  [2, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "TTCCATCCTCTTTGCCGGAAGG  [1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "TTCCATCCTCTTTGTGTTACCG  [4, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "\n",
       "                       Maxima  Maxima-Count  nucleosomal-score  \n",
       "AAACTACCAGAAACCCGAGATA    [0]           1.0         145.544305  \n",
       "AAACTACCAGAAACCTAAGTGG    [0]           1.0         102.513844  \n",
       "AAACTACCAGAAACGGATCAGT    [0]           1.0         140.045502  \n",
       "AAACTACCAGAAACGTCCCGTT    [0]           1.0          19.623836  \n",
       "AAACTACCAGAAACTAGCCCTA    [0]           1.0         152.003069  \n",
       "...                       ...           ...                ...  \n",
       "TTCCATCCTCTTTCGCGTGTAA    [0]           1.0         220.573902  \n",
       "TTCCATCCTCTTTCTGCAGACT    [0]           1.0         324.955404  \n",
       "TTCCATCCTCTTTCTGGCGCAG    [0]           1.0         324.955404  \n",
       "TTCCATCCTCTTTGCCGGAAGG     []           0.0                inf  \n",
       "TTCCATCCTCTTTGTGTTACCG    [0]           1.0         220.573902  \n",
       "\n",
       "[73652 rows x 20 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d36a2d65",
   "metadata": {},
   "outputs": [],
   "source": [
    "#if (qc_columns['mean_insertsize'] is None or qc_columns['insertsize_count'] is None) and calc_mean_insertsize:\n",
    "#    \n",
    "#    if use_bam:\n",
    "#        atac.add_insertsize(adata, bam=bam_file, barcode_tag=barcode_tag)     \n",
    "#    else:\n",
    "#        atac.add_insertsize(adata, fragments=fragments_file, barcode_tag=barcode_tag)\n",
    "#        \n",
    "#    # update column dictionary\n",
    "#    qc_columns['mean_insertsize'] = 'mean_insertsize'\n",
    "#    qc_columns['insertsize_count'] = 'insertsize_count'\n",
    "#    \n",
    "#    # update infoprocess table\n",
    "#    build_legend(adata, 'mean_insertsize', 'mean_insertsize')\n",
    "#    build_legend(adata, 'insertsize_count', 'insertsize_count')\n",
    "#    \n",
    "#    atac.plot_insertsize(adata, barcodes=None)\n",
    "#    \n",
    "#adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a1bfbd4",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### 3. promotor enrichment "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c414cbec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converting GTF to BED...\n",
      "Converting BAM to fragments file! This may take a while...\n",
      "Finished creating fragments file. Now sorting...\n",
      "Finished sorting fragments\n",
      "Finding overlaps...\n",
      "Calculating percentage...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e05106fd7f1446cb90fb99f37505f298",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Progress:   0%|          | 0/15 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4dfafe5df5c64da78822f51e2d29949e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Progress:   0%|          | 0/76 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Adding results to adata object...\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "if qc_columns['n_fragments_in_promoters'] is None and calc_promotor_enrichment:\n",
    "\n",
    "    if use_bam:\n",
    "        overlap.pct_fragments_in_promoters(adata, gtf_file=promoters_gtf, species=species, bam_file=bam_file, cb_col=None, nproc=1, cb_tag=barcode_tag)\n",
    "    else:\n",
    "        overlap.pct_fragments_in_promoters(adata, gtf_file=promoters_gtf, species=species, fragments_file=fragments_file, cb_col=None, nproc=1, cb_tag=barcode_tag)\n",
    "\n",
    "    # n_total_fragments, n_fragments_in_promoters, pct_fragments_in_promoters\n",
    "    # update column dictionary\n",
    "    qc_columns['n_total_fragments'] = 'n_total_fragments'\n",
    "    qc_columns['n_fragments_in_promoters'] = 'n_fragments_in_promoters'\n",
    "    qc_columns['pct_fragments_in_promoters'] = 'pct_fragments_in_promoters'\n",
    "    # update infoprocess table\n",
    "    build_legend(adata, 'n_total_fragments', 'n_total_fragments')\n",
    "    build_legend(adata, 'n_fragments_in_promoters', 'n_fragments_in_promoters')\n",
    "    build_legend(adata, 'pct_fragments_in_promoters', 'pct_fragments_in_promoters')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dae390c8",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## Save adata to .h5ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "bee812e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/mnt/workspace/jdetlef/processed_data/Esophagus_146_0.01/assembling/anndata/Esophagus_146_0.01.h5ad'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_output = tree.assembled_anndata\n",
    "adata_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4c65b408",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TN</th>\n",
       "      <th>UM</th>\n",
       "      <th>PP</th>\n",
       "      <th>UQ</th>\n",
       "      <th>CM</th>\n",
       "      <th>file</th>\n",
       "      <th>sample</th>\n",
       "      <th>Sample_new</th>\n",
       "      <th>n_features_by_counts</th>\n",
       "      <th>log1p_n_features_by_counts</th>\n",
       "      <th>...</th>\n",
       "      <th>Median-Fragment-Length</th>\n",
       "      <th>Fragments</th>\n",
       "      <th>Fragment-Count</th>\n",
       "      <th>Distribution</th>\n",
       "      <th>Maxima</th>\n",
       "      <th>Maxima-Count</th>\n",
       "      <th>nucleosomal-score</th>\n",
       "      <th>n_fragments_in_promoters</th>\n",
       "      <th>n_total_fragments</th>\n",
       "      <th>pct_fragments_in_promoters</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACCCGAGATA</th>\n",
       "      <td>33.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>5</td>\n",
       "      <td>1.791759</td>\n",
       "      <td>...</td>\n",
       "      <td>49</td>\n",
       "      <td>[123, 46, 90, 49, 201, 61, 178, 44, 34, 42, 38]</td>\n",
       "      <td>11.0</td>\n",
       "      <td>[9, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>145.544305</td>\n",
       "      <td>4.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACCTAAGTGG</th>\n",
       "      <td>52.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>17</td>\n",
       "      <td>2.890372</td>\n",
       "      <td>...</td>\n",
       "      <td>138.5</td>\n",
       "      <td>[72, 417, 148, 40, 36, 129, 207, 89, 166, 111,...</td>\n",
       "      <td>26.0</td>\n",
       "      <td>[16, 7, 2, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...</td>\n",
       "      <td>[0]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>102.513844</td>\n",
       "      <td>5.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0.178571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACGGATCAGT</th>\n",
       "      <td>27.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>10</td>\n",
       "      <td>2.397895</td>\n",
       "      <td>...</td>\n",
       "      <td>74.0</td>\n",
       "      <td>[407, 78, 50, 52, 26, 70, 81, 178, 496, 28, 53...</td>\n",
       "      <td>12.0</td>\n",
       "      <td>[8, 1, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>140.045502</td>\n",
       "      <td>5.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.263158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACGTCCCGTT</th>\n",
       "      <td>6223.0</td>\n",
       "      <td>5231.0</td>\n",
       "      <td>5213.0</td>\n",
       "      <td>3779.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>1648</td>\n",
       "      <td>7.407924</td>\n",
       "      <td>...</td>\n",
       "      <td>106</td>\n",
       "      <td>[60, 54, 31, 201, 41, 267, 49, 259, 275, 77, 6...</td>\n",
       "      <td>3475.0</td>\n",
       "      <td>[2351, 871, 229, 20, 4, 0, 0, 0, 0, 0, 0, 0, 0...</td>\n",
       "      <td>[0]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>19.623836</td>\n",
       "      <td>1296.0</td>\n",
       "      <td>5264.0</td>\n",
       "      <td>0.246201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACTAGCCCTA</th>\n",
       "      <td>41.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>13</td>\n",
       "      <td>2.639057</td>\n",
       "      <td>...</td>\n",
       "      <td>131.5</td>\n",
       "      <td>[56, 350, 37, 391, 212, 40, 207, 48, 49, 215]</td>\n",
       "      <td>10.0</td>\n",
       "      <td>[5, 3, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>152.003069</td>\n",
       "      <td>6.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.315789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTCGCGTGTAA</th>\n",
       "      <td>9.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>3</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>...</td>\n",
       "      <td>94</td>\n",
       "      <td>[91, 94, 199, 77, 168]</td>\n",
       "      <td>5.0</td>\n",
       "      <td>[4, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>220.573902</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTCTGCAGACT</th>\n",
       "      <td>10.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>3</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>...</td>\n",
       "      <td>58</td>\n",
       "      <td>[58, 147, 56]</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>324.955404</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.428571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTCTGGCGCAG</th>\n",
       "      <td>17.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>46</td>\n",
       "      <td>[176, 36, 46]</td>\n",
       "      <td>3.0</td>\n",
       "      <td>[2, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>324.955404</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.444444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTGCCGGAAGG</th>\n",
       "      <td>8.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>5</td>\n",
       "      <td>1.791759</td>\n",
       "      <td>...</td>\n",
       "      <td>163.5</td>\n",
       "      <td>[266, 61]</td>\n",
       "      <td>2.0</td>\n",
       "      <td>[1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTGTGTTACCG</th>\n",
       "      <td>19.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>2</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>...</td>\n",
       "      <td>100</td>\n",
       "      <td>[122, 46, 241, 100, 66]</td>\n",
       "      <td>5.0</td>\n",
       "      <td>[4, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0]</td>\n",
       "      <td>1.0</td>\n",
       "      <td>220.573902</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>73652 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            TN      UM      PP      UQ    CM  \\\n",
       "AAACTACCAGAAACCCGAGATA    33.0    18.0    18.0    15.0   0.0   \n",
       "AAACTACCAGAAACCTAAGTGG    52.0    36.0    35.0    32.0   1.0   \n",
       "AAACTACCAGAAACGGATCAGT    27.0    19.0    19.0    19.0   0.0   \n",
       "AAACTACCAGAAACGTCCCGTT  6223.0  5231.0  5213.0  3779.0  18.0   \n",
       "AAACTACCAGAAACTAGCCCTA    41.0    29.0    29.0    26.0   0.0   \n",
       "...                        ...     ...     ...     ...   ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA     9.0     6.0     6.0     6.0   0.0   \n",
       "TTCCATCCTCTTTCTGCAGACT    10.0     6.0     6.0     6.0   1.0   \n",
       "TTCCATCCTCTTTCTGGCGCAG    17.0     5.0     5.0     5.0   1.0   \n",
       "TTCCATCCTCTTTGCCGGAAGG     8.0     7.0     7.0     7.0   0.0   \n",
       "TTCCATCCTCTTTGTGTTACCG    19.0     9.0     9.0     7.0   0.0   \n",
       "\n",
       "                                                                     file  \\\n",
       "AAACTACCAGAAACCCGAGATA  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "AAACTACCAGAAACCTAAGTGG  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "AAACTACCAGAAACGGATCAGT  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "AAACTACCAGAAACGTCCCGTT  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "AAACTACCAGAAACTAGCCCTA  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "...                                                                   ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "TTCCATCCTCTTTCTGCAGACT  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "TTCCATCCTCTTTCTGGCGCAG  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "TTCCATCCTCTTTGCCGGAAGG  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "TTCCATCCTCTTTGTGTTACCG  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "\n",
       "                       sample Sample_new  n_features_by_counts  \\\n",
       "AAACTACCAGAAACCCGAGATA     MB    sample1                     5   \n",
       "AAACTACCAGAAACCTAAGTGG     MB    sample1                    17   \n",
       "AAACTACCAGAAACGGATCAGT     MB    sample1                    10   \n",
       "AAACTACCAGAAACGTCCCGTT     MB    sample1                  1648   \n",
       "AAACTACCAGAAACTAGCCCTA     MB    sample1                    13   \n",
       "...                       ...        ...                   ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA     MB    sample1                     3   \n",
       "TTCCATCCTCTTTCTGCAGACT     MB    sample1                     3   \n",
       "TTCCATCCTCTTTCTGGCGCAG     MB    sample1                     0   \n",
       "TTCCATCCTCTTTGCCGGAAGG     MB    sample1                     5   \n",
       "TTCCATCCTCTTTGTGTTACCG     MB    sample1                     2   \n",
       "\n",
       "                        log1p_n_features_by_counts  ...  \\\n",
       "AAACTACCAGAAACCCGAGATA                    1.791759  ...   \n",
       "AAACTACCAGAAACCTAAGTGG                    2.890372  ...   \n",
       "AAACTACCAGAAACGGATCAGT                    2.397895  ...   \n",
       "AAACTACCAGAAACGTCCCGTT                    7.407924  ...   \n",
       "AAACTACCAGAAACTAGCCCTA                    2.639057  ...   \n",
       "...                                            ...  ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA                    1.386294  ...   \n",
       "TTCCATCCTCTTTCTGCAGACT                    1.386294  ...   \n",
       "TTCCATCCTCTTTCTGGCGCAG                    0.000000  ...   \n",
       "TTCCATCCTCTTTGCCGGAAGG                    1.791759  ...   \n",
       "TTCCATCCTCTTTGTGTTACCG                    1.098612  ...   \n",
       "\n",
       "                        Median-Fragment-Length  \\\n",
       "AAACTACCAGAAACCCGAGATA                      49   \n",
       "AAACTACCAGAAACCTAAGTGG                   138.5   \n",
       "AAACTACCAGAAACGGATCAGT                    74.0   \n",
       "AAACTACCAGAAACGTCCCGTT                     106   \n",
       "AAACTACCAGAAACTAGCCCTA                   131.5   \n",
       "...                                        ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA                      94   \n",
       "TTCCATCCTCTTTCTGCAGACT                      58   \n",
       "TTCCATCCTCTTTCTGGCGCAG                      46   \n",
       "TTCCATCCTCTTTGCCGGAAGG                   163.5   \n",
       "TTCCATCCTCTTTGTGTTACCG                     100   \n",
       "\n",
       "                                                                Fragments  \\\n",
       "AAACTACCAGAAACCCGAGATA    [123, 46, 90, 49, 201, 61, 178, 44, 34, 42, 38]   \n",
       "AAACTACCAGAAACCTAAGTGG  [72, 417, 148, 40, 36, 129, 207, 89, 166, 111,...   \n",
       "AAACTACCAGAAACGGATCAGT  [407, 78, 50, 52, 26, 70, 81, 178, 496, 28, 53...   \n",
       "AAACTACCAGAAACGTCCCGTT  [60, 54, 31, 201, 41, 267, 49, 259, 275, 77, 6...   \n",
       "AAACTACCAGAAACTAGCCCTA      [56, 350, 37, 391, 212, 40, 207, 48, 49, 215]   \n",
       "...                                                                   ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA                             [91, 94, 199, 77, 168]   \n",
       "TTCCATCCTCTTTCTGCAGACT                                      [58, 147, 56]   \n",
       "TTCCATCCTCTTTCTGGCGCAG                                      [176, 36, 46]   \n",
       "TTCCATCCTCTTTGCCGGAAGG                                          [266, 61]   \n",
       "TTCCATCCTCTTTGTGTTACCG                            [122, 46, 241, 100, 66]   \n",
       "\n",
       "                       Fragment-Count  \\\n",
       "AAACTACCAGAAACCCGAGATA           11.0   \n",
       "AAACTACCAGAAACCTAAGTGG           26.0   \n",
       "AAACTACCAGAAACGGATCAGT           12.0   \n",
       "AAACTACCAGAAACGTCCCGTT         3475.0   \n",
       "AAACTACCAGAAACTAGCCCTA           10.0   \n",
       "...                               ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA            5.0   \n",
       "TTCCATCCTCTTTCTGCAGACT            3.0   \n",
       "TTCCATCCTCTTTCTGGCGCAG            3.0   \n",
       "TTCCATCCTCTTTGCCGGAAGG            2.0   \n",
       "TTCCATCCTCTTTGTGTTACCG            5.0   \n",
       "\n",
       "                                                             Distribution  \\\n",
       "AAACTACCAGAAACCCGAGATA  [9, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "AAACTACCAGAAACCTAAGTGG  [16, 7, 2, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...   \n",
       "AAACTACCAGAAACGGATCAGT  [8, 1, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "AAACTACCAGAAACGTCCCGTT  [2351, 871, 229, 20, 4, 0, 0, 0, 0, 0, 0, 0, 0...   \n",
       "AAACTACCAGAAACTAGCCCTA  [5, 3, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "...                                                                   ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA  [4, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "TTCCATCCTCTTTCTGCAGACT  [3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "TTCCATCCTCTTTCTGGCGCAG  [2, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "TTCCATCCTCTTTGCCGGAAGG  [1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "TTCCATCCTCTTTGTGTTACCG  [4, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "\n",
       "                       Maxima  Maxima-Count nucleosomal-score  \\\n",
       "AAACTACCAGAAACCCGAGATA    [0]           1.0        145.544305   \n",
       "AAACTACCAGAAACCTAAGTGG    [0]           1.0        102.513844   \n",
       "AAACTACCAGAAACGGATCAGT    [0]           1.0        140.045502   \n",
       "AAACTACCAGAAACGTCCCGTT    [0]           1.0         19.623836   \n",
       "AAACTACCAGAAACTAGCCCTA    [0]           1.0        152.003069   \n",
       "...                       ...           ...               ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA    [0]           1.0        220.573902   \n",
       "TTCCATCCTCTTTCTGCAGACT    [0]           1.0        324.955404   \n",
       "TTCCATCCTCTTTCTGGCGCAG    [0]           1.0        324.955404   \n",
       "TTCCATCCTCTTTGCCGGAAGG     []           0.0               inf   \n",
       "TTCCATCCTCTTTGTGTTACCG    [0]           1.0        220.573902   \n",
       "\n",
       "                       n_fragments_in_promoters  n_total_fragments  \\\n",
       "AAACTACCAGAAACCCGAGATA                      4.0               16.0   \n",
       "AAACTACCAGAAACCTAAGTGG                      5.0               28.0   \n",
       "AAACTACCAGAAACGGATCAGT                      5.0               19.0   \n",
       "AAACTACCAGAAACGTCCCGTT                   1296.0             5264.0   \n",
       "AAACTACCAGAAACTAGCCCTA                      6.0               19.0   \n",
       "...                                         ...                ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA                      1.0                6.0   \n",
       "TTCCATCCTCTTTCTGCAGACT                      3.0                7.0   \n",
       "TTCCATCCTCTTTCTGGCGCAG                      4.0                9.0   \n",
       "TTCCATCCTCTTTGCCGGAAGG                      1.0                5.0   \n",
       "TTCCATCCTCTTTGTGTTACCG                      0.0                8.0   \n",
       "\n",
       "                        pct_fragments_in_promoters  \n",
       "AAACTACCAGAAACCCGAGATA                    0.250000  \n",
       "AAACTACCAGAAACCTAAGTGG                    0.178571  \n",
       "AAACTACCAGAAACGGATCAGT                    0.263158  \n",
       "AAACTACCAGAAACGTCCCGTT                    0.246201  \n",
       "AAACTACCAGAAACTAGCCCTA                    0.315789  \n",
       "...                                            ...  \n",
       "TTCCATCCTCTTTCGCGTGTAA                    0.166667  \n",
       "TTCCATCCTCTTTCTGCAGACT                    0.428571  \n",
       "TTCCATCCTCTTTCTGGCGCAG                    0.444444  \n",
       "TTCCATCCTCTTTGCCGGAAGG                    0.200000  \n",
       "TTCCATCCTCTTTGTGTTACCG                    0.000000  \n",
       "\n",
       "[73652 rows x 23 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "71e54842",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['Fragments'] = adata.obs['Fragments'].astype(str)\n",
    "adata.obs['Distribution'] = adata.obs['Distribution'].astype(str)\n",
    "#adata.obs['Distribution'] = adata.obs['Distribution'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "5c00ab53",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['Maxima'] = adata.obs['Maxima'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f3c0ad8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAACTACCAGAAACCCGAGATA      33.0\n",
      "AAACTACCAGAAACCTAAGTGG      52.0\n",
      "AAACTACCAGAAACGGATCAGT      27.0\n",
      "AAACTACCAGAAACGTCCCGTT    6223.0\n",
      "AAACTACCAGAAACTAGCCCTA      41.0\n",
      "                           ...  \n",
      "TTCCATCCTCTTTCGCGTGTAA       9.0\n",
      "TTCCATCCTCTTTCTGCAGACT      10.0\n",
      "TTCCATCCTCTTTCTGGCGCAG      17.0\n",
      "TTCCATCCTCTTTGCCGGAAGG       8.0\n",
      "TTCCATCCTCTTTGTGTTACCG      19.0\n",
      "Name: TN, Length: 73652, dtype: float64\n",
      "AAACTACCAGAAACCCGAGATA      18.0\n",
      "AAACTACCAGAAACCTAAGTGG      36.0\n",
      "AAACTACCAGAAACGGATCAGT      19.0\n",
      "AAACTACCAGAAACGTCCCGTT    5231.0\n",
      "AAACTACCAGAAACTAGCCCTA      29.0\n",
      "                           ...  \n",
      "TTCCATCCTCTTTCGCGTGTAA       6.0\n",
      "TTCCATCCTCTTTCTGCAGACT       6.0\n",
      "TTCCATCCTCTTTCTGGCGCAG       5.0\n",
      "TTCCATCCTCTTTGCCGGAAGG       7.0\n",
      "TTCCATCCTCTTTGTGTTACCG       9.0\n",
      "Name: UM, Length: 73652, dtype: float64\n",
      "AAACTACCAGAAACCCGAGATA      18.0\n",
      "AAACTACCAGAAACCTAAGTGG      35.0\n",
      "AAACTACCAGAAACGGATCAGT      19.0\n",
      "AAACTACCAGAAACGTCCCGTT    5213.0\n",
      "AAACTACCAGAAACTAGCCCTA      29.0\n",
      "                           ...  \n",
      "TTCCATCCTCTTTCGCGTGTAA       6.0\n",
      "TTCCATCCTCTTTCTGCAGACT       6.0\n",
      "TTCCATCCTCTTTCTGGCGCAG       5.0\n",
      "TTCCATCCTCTTTGCCGGAAGG       7.0\n",
      "TTCCATCCTCTTTGTGTTACCG       9.0\n",
      "Name: PP, Length: 73652, dtype: float64\n",
      "AAACTACCAGAAACCCGAGATA      15.0\n",
      "AAACTACCAGAAACCTAAGTGG      32.0\n",
      "AAACTACCAGAAACGGATCAGT      19.0\n",
      "AAACTACCAGAAACGTCCCGTT    3779.0\n",
      "AAACTACCAGAAACTAGCCCTA      26.0\n",
      "                           ...  \n",
      "TTCCATCCTCTTTCGCGTGTAA       6.0\n",
      "TTCCATCCTCTTTCTGCAGACT       6.0\n",
      "TTCCATCCTCTTTCTGGCGCAG       5.0\n",
      "TTCCATCCTCTTTGCCGGAAGG       7.0\n",
      "TTCCATCCTCTTTGTGTTACCG       7.0\n",
      "Name: UQ, Length: 73652, dtype: float64\n",
      "AAACTACCAGAAACCCGAGATA     0.0\n",
      "AAACTACCAGAAACCTAAGTGG     1.0\n",
      "AAACTACCAGAAACGGATCAGT     0.0\n",
      "AAACTACCAGAAACGTCCCGTT    18.0\n",
      "AAACTACCAGAAACTAGCCCTA     0.0\n",
      "                          ... \n",
      "TTCCATCCTCTTTCGCGTGTAA     0.0\n",
      "TTCCATCCTCTTTCTGCAGACT     1.0\n",
      "TTCCATCCTCTTTCTGGCGCAG     1.0\n",
      "TTCCATCCTCTTTGCCGGAAGG     0.0\n",
      "TTCCATCCTCTTTGTGTTACCG     0.0\n",
      "Name: CM, Length: 73652, dtype: float64\n",
      "AAACTACCAGAAACCCGAGATA    /mnt/workspace/jdetlef/data/anndata/esophagus_...\n",
      "AAACTACCAGAAACCTAAGTGG    /mnt/workspace/jdetlef/data/anndata/esophagus_...\n",
      "AAACTACCAGAAACGGATCAGT    /mnt/workspace/jdetlef/data/anndata/esophagus_...\n",
      "AAACTACCAGAAACGTCCCGTT    /mnt/workspace/jdetlef/data/anndata/esophagus_...\n",
      "AAACTACCAGAAACTAGCCCTA    /mnt/workspace/jdetlef/data/anndata/esophagus_...\n",
      "                                                ...                        \n",
      "TTCCATCCTCTTTCGCGTGTAA    /mnt/workspace/jdetlef/data/anndata/esophagus_...\n",
      "TTCCATCCTCTTTCTGCAGACT    /mnt/workspace/jdetlef/data/anndata/esophagus_...\n",
      "TTCCATCCTCTTTCTGGCGCAG    /mnt/workspace/jdetlef/data/anndata/esophagus_...\n",
      "TTCCATCCTCTTTGCCGGAAGG    /mnt/workspace/jdetlef/data/anndata/esophagus_...\n",
      "TTCCATCCTCTTTGTGTTACCG    /mnt/workspace/jdetlef/data/anndata/esophagus_...\n",
      "Name: file, Length: 73652, dtype: category\n",
      "Categories (1, object): ['/mnt/workspace/jdetlef/data/anndata/esophagus...]\n",
      "AAACTACCAGAAACCCGAGATA    MB\n",
      "AAACTACCAGAAACCTAAGTGG    MB\n",
      "AAACTACCAGAAACGGATCAGT    MB\n",
      "AAACTACCAGAAACGTCCCGTT    MB\n",
      "AAACTACCAGAAACTAGCCCTA    MB\n",
      "                          ..\n",
      "TTCCATCCTCTTTCGCGTGTAA    MB\n",
      "TTCCATCCTCTTTCTGCAGACT    MB\n",
      "TTCCATCCTCTTTCTGGCGCAG    MB\n",
      "TTCCATCCTCTTTGCCGGAAGG    MB\n",
      "TTCCATCCTCTTTGTGTTACCG    MB\n",
      "Name: sample, Length: 73652, dtype: category\n",
      "Categories (1, object): ['MB']\n",
      "AAACTACCAGAAACCCGAGATA    sample1\n",
      "AAACTACCAGAAACCTAAGTGG    sample1\n",
      "AAACTACCAGAAACGGATCAGT    sample1\n",
      "AAACTACCAGAAACGTCCCGTT    sample1\n",
      "AAACTACCAGAAACTAGCCCTA    sample1\n",
      "                           ...   \n",
      "TTCCATCCTCTTTCGCGTGTAA    sample1\n",
      "TTCCATCCTCTTTCTGCAGACT    sample1\n",
      "TTCCATCCTCTTTCTGGCGCAG    sample1\n",
      "TTCCATCCTCTTTGCCGGAAGG    sample1\n",
      "TTCCATCCTCTTTGTGTTACCG    sample1\n",
      "Name: Sample_new, Length: 73652, dtype: category\n",
      "Categories (1, object): ['sample1']\n",
      "AAACTACCAGAAACCCGAGATA       5\n",
      "AAACTACCAGAAACCTAAGTGG      17\n",
      "AAACTACCAGAAACGGATCAGT      10\n",
      "AAACTACCAGAAACGTCCCGTT    1648\n",
      "AAACTACCAGAAACTAGCCCTA      13\n",
      "                          ... \n",
      "TTCCATCCTCTTTCGCGTGTAA       3\n",
      "TTCCATCCTCTTTCTGCAGACT       3\n",
      "TTCCATCCTCTTTCTGGCGCAG       0\n",
      "TTCCATCCTCTTTGCCGGAAGG       5\n",
      "TTCCATCCTCTTTGTGTTACCG       2\n",
      "Name: n_features_by_counts, Length: 73652, dtype: int32\n",
      "AAACTACCAGAAACCCGAGATA    1.791759\n",
      "AAACTACCAGAAACCTAAGTGG    2.890372\n",
      "AAACTACCAGAAACGGATCAGT    2.397895\n",
      "AAACTACCAGAAACGTCCCGTT    7.407924\n",
      "AAACTACCAGAAACTAGCCCTA    2.639057\n",
      "                            ...   \n",
      "TTCCATCCTCTTTCGCGTGTAA    1.386294\n",
      "TTCCATCCTCTTTCTGCAGACT    1.386294\n",
      "TTCCATCCTCTTTCTGGCGCAG    0.000000\n",
      "TTCCATCCTCTTTGCCGGAAGG    1.791759\n",
      "TTCCATCCTCTTTGTGTTACCG    1.098612\n",
      "Name: log1p_n_features_by_counts, Length: 73652, dtype: float64\n",
      "AAACTACCAGAAACCCGAGATA       5.0\n",
      "AAACTACCAGAAACCTAAGTGG      17.0\n",
      "AAACTACCAGAAACGGATCAGT      10.0\n",
      "AAACTACCAGAAACGTCCCGTT    1791.0\n",
      "AAACTACCAGAAACTAGCCCTA      13.0\n",
      "                           ...  \n",
      "TTCCATCCTCTTTCGCGTGTAA       3.0\n",
      "TTCCATCCTCTTTCTGCAGACT       3.0\n",
      "TTCCATCCTCTTTCTGGCGCAG       0.0\n",
      "TTCCATCCTCTTTGCCGGAAGG       5.0\n",
      "TTCCATCCTCTTTGTGTTACCG       2.0\n",
      "Name: total_counts, Length: 73652, dtype: float32\n",
      "AAACTACCAGAAACCCGAGATA    1.791759\n",
      "AAACTACCAGAAACCTAAGTGG    2.890372\n",
      "AAACTACCAGAAACGGATCAGT    2.397895\n",
      "AAACTACCAGAAACGTCCCGTT    7.491087\n",
      "AAACTACCAGAAACTAGCCCTA    2.639057\n",
      "                            ...   \n",
      "TTCCATCCTCTTTCGCGTGTAA    1.386294\n",
      "TTCCATCCTCTTTCTGCAGACT    1.386294\n",
      "TTCCATCCTCTTTCTGGCGCAG    0.000000\n",
      "TTCCATCCTCTTTGCCGGAAGG    1.791759\n",
      "TTCCATCCTCTTTGTGTTACCG    1.098612\n",
      "Name: log1p_total_counts, Length: 73652, dtype: float32\n",
      "AAACTACCAGAAACCCGAGATA     82.36\n",
      "AAACTACCAGAAACCTAAGTGG    170.38\n",
      "AAACTACCAGAAACGGATCAGT    161.17\n",
      "AAACTACCAGAAACGTCCCGTT    144.03\n",
      "AAACTACCAGAAACTAGCCCTA     160.5\n",
      "                           ...  \n",
      "TTCCATCCTCTTTCGCGTGTAA     125.8\n",
      "TTCCATCCTCTTTCTGCAGACT      87.0\n",
      "TTCCATCCTCTTTCTGGCGCAG      86.0\n",
      "TTCCATCCTCTTTGCCGGAAGG     163.5\n",
      "TTCCATCCTCTTTGTGTTACCG     115.0\n",
      "Name: Mean-Fragment-Length, Length: 73652, dtype: object\n",
      "AAACTACCAGAAACCCGAGATA       49\n",
      "AAACTACCAGAAACCTAAGTGG    138.5\n",
      "AAACTACCAGAAACGGATCAGT     74.0\n",
      "AAACTACCAGAAACGTCCCGTT      106\n",
      "AAACTACCAGAAACTAGCCCTA    131.5\n",
      "                          ...  \n",
      "TTCCATCCTCTTTCGCGTGTAA       94\n",
      "TTCCATCCTCTTTCTGCAGACT       58\n",
      "TTCCATCCTCTTTCTGGCGCAG       46\n",
      "TTCCATCCTCTTTGCCGGAAGG    163.5\n",
      "TTCCATCCTCTTTGTGTTACCG      100\n",
      "Name: Median-Fragment-Length, Length: 73652, dtype: object\n",
      "AAACTACCAGAAACCCGAGATA      [123, 46, 90, 49, 201, 61, 178, 44, 34, 42, 38]\n",
      "AAACTACCAGAAACCTAAGTGG    [72, 417, 148, 40, 36, 129, 207, 89, 166, 111,...\n",
      "AAACTACCAGAAACGGATCAGT    [407, 78, 50, 52, 26, 70, 81, 178, 496, 28, 53...\n",
      "AAACTACCAGAAACGTCCCGTT    [60, 54, 31, 201, 41, 267, 49, 259, 275, 77, 6...\n",
      "AAACTACCAGAAACTAGCCCTA        [56, 350, 37, 391, 212, 40, 207, 48, 49, 215]\n",
      "                                                ...                        \n",
      "TTCCATCCTCTTTCGCGTGTAA                               [91, 94, 199, 77, 168]\n",
      "TTCCATCCTCTTTCTGCAGACT                                        [58, 147, 56]\n",
      "TTCCATCCTCTTTCTGGCGCAG                                        [176, 36, 46]\n",
      "TTCCATCCTCTTTGCCGGAAGG                                            [266, 61]\n",
      "TTCCATCCTCTTTGTGTTACCG                              [122, 46, 241, 100, 66]\n",
      "Name: Fragments, Length: 73652, dtype: category\n",
      "Categories (71813, object): ['[10, 37, 37, 212]', '[10, 38, 71, 42]', '[10, 54, 219, 186, 229, 155, 89, 33]', '[10, 55, 35, 262, 41, 31, 56, 147, 93, 30, 18..., ..., '[992, 98, 50, 256, 203, 79, 46, 30, 200, 167]', '[1036, 73, 70, 37, 129, 49, 50, 52, 135, 88]', '[1737, 67, 200, 354, 76, 49, 47, 49, 107, 292]', 'nan']\n",
      "AAACTACCAGAAACCCGAGATA      11.0\n",
      "AAACTACCAGAAACCTAAGTGG      26.0\n",
      "AAACTACCAGAAACGGATCAGT      12.0\n",
      "AAACTACCAGAAACGTCCCGTT    3475.0\n",
      "AAACTACCAGAAACTAGCCCTA      10.0\n",
      "                           ...  \n",
      "TTCCATCCTCTTTCGCGTGTAA       5.0\n",
      "TTCCATCCTCTTTCTGCAGACT       3.0\n",
      "TTCCATCCTCTTTCTGGCGCAG       3.0\n",
      "TTCCATCCTCTTTGCCGGAAGG       2.0\n",
      "TTCCATCCTCTTTGTGTTACCG       5.0\n",
      "Name: Fragment-Count, Length: 73652, dtype: float64\n",
      "AAACTACCAGAAACCCGAGATA    [9, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...\n",
      "AAACTACCAGAAACCTAAGTGG    [16, 7, 2, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,...\n",
      "AAACTACCAGAAACGGATCAGT    [8, 1, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...\n",
      "AAACTACCAGAAACGTCCCGTT    [2351, 871, 229, 20, 4, 0, 0, 0, 0, 0, 0, 0, 0...\n",
      "AAACTACCAGAAACTAGCCCTA    [5, 3, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...\n",
      "                                                ...                        \n",
      "TTCCATCCTCTTTCGCGTGTAA    [4, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...\n",
      "TTCCATCCTCTTTCTGCAGACT    [3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...\n",
      "TTCCATCCTCTTTCTGGCGCAG    [2, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...\n",
      "TTCCATCCTCTTTGCCGGAAGG    [1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...\n",
      "TTCCATCCTCTTTGTGTTACCG    [4, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...\n",
      "Name: Distribution, Length: 73652, dtype: category\n",
      "Categories (16989, object): ['[0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,..., '[0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,..., '[0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,..., '[0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,..., ..., '[65122, 25076, 6586, 630, 39, 8, 1, 1, 0, 0, ..., '[348826, 111420, 25716, 2379, 185, 38, 11, 3,..., '[381738, 117920, 26581, 2367, 165, 35, 16, 3,..., 'nan']\n",
      "AAACTACCAGAAACCCGAGATA    [0]\n",
      "AAACTACCAGAAACCTAAGTGG    [0]\n",
      "AAACTACCAGAAACGGATCAGT    [0]\n",
      "AAACTACCAGAAACGTCCCGTT    [0]\n",
      "AAACTACCAGAAACTAGCCCTA    [0]\n",
      "                         ... \n",
      "TTCCATCCTCTTTCGCGTGTAA    [0]\n",
      "TTCCATCCTCTTTCTGCAGACT    [0]\n",
      "TTCCATCCTCTTTCTGGCGCAG    [0]\n",
      "TTCCATCCTCTTTGCCGGAAGG     []\n",
      "TTCCATCCTCTTTGTGTTACCG    [0]\n",
      "Name: Maxima, Length: 73652, dtype: category\n",
      "Categories (89, object): ['[0, 3]', '[0, 4]', '[0, 5]', '[0, 6, 9]', ..., '[6]', '[9]', '[]', 'nan']\n",
      "AAACTACCAGAAACCCGAGATA    1.0\n",
      "AAACTACCAGAAACCTAAGTGG    1.0\n",
      "AAACTACCAGAAACGGATCAGT    1.0\n",
      "AAACTACCAGAAACGTCCCGTT    1.0\n",
      "AAACTACCAGAAACTAGCCCTA    1.0\n",
      "                         ... \n",
      "TTCCATCCTCTTTCGCGTGTAA    1.0\n",
      "TTCCATCCTCTTTCTGCAGACT    1.0\n",
      "TTCCATCCTCTTTCTGGCGCAG    1.0\n",
      "TTCCATCCTCTTTGCCGGAAGG    0.0\n",
      "TTCCATCCTCTTTGTGTTACCG    1.0\n",
      "Name: Maxima-Count, Length: 73652, dtype: float64\n",
      "AAACTACCAGAAACCCGAGATA    145.544305\n",
      "AAACTACCAGAAACCTAAGTGG    102.513844\n",
      "AAACTACCAGAAACGGATCAGT    140.045502\n",
      "AAACTACCAGAAACGTCCCGTT     19.623836\n",
      "AAACTACCAGAAACTAGCCCTA    152.003069\n",
      "                             ...    \n",
      "TTCCATCCTCTTTCGCGTGTAA    220.573902\n",
      "TTCCATCCTCTTTCTGCAGACT    324.955404\n",
      "TTCCATCCTCTTTCTGGCGCAG    324.955404\n",
      "TTCCATCCTCTTTGCCGGAAGG           inf\n",
      "TTCCATCCTCTTTGTGTTACCG    220.573902\n",
      "Name: nucleosomal-score, Length: 73652, dtype: float64\n",
      "AAACTACCAGAAACCCGAGATA       4.0\n",
      "AAACTACCAGAAACCTAAGTGG       5.0\n",
      "AAACTACCAGAAACGGATCAGT       5.0\n",
      "AAACTACCAGAAACGTCCCGTT    1296.0\n",
      "AAACTACCAGAAACTAGCCCTA       6.0\n",
      "                           ...  \n",
      "TTCCATCCTCTTTCGCGTGTAA       1.0\n",
      "TTCCATCCTCTTTCTGCAGACT       3.0\n",
      "TTCCATCCTCTTTCTGGCGCAG       4.0\n",
      "TTCCATCCTCTTTGCCGGAAGG       1.0\n",
      "TTCCATCCTCTTTGTGTTACCG       0.0\n",
      "Name: n_fragments_in_promoters, Length: 73652, dtype: float64\n",
      "AAACTACCAGAAACCCGAGATA      16.0\n",
      "AAACTACCAGAAACCTAAGTGG      28.0\n",
      "AAACTACCAGAAACGGATCAGT      19.0\n",
      "AAACTACCAGAAACGTCCCGTT    5264.0\n",
      "AAACTACCAGAAACTAGCCCTA      19.0\n",
      "                           ...  \n",
      "TTCCATCCTCTTTCGCGTGTAA       6.0\n",
      "TTCCATCCTCTTTCTGCAGACT       7.0\n",
      "TTCCATCCTCTTTCTGGCGCAG       9.0\n",
      "TTCCATCCTCTTTGCCGGAAGG       5.0\n",
      "TTCCATCCTCTTTGTGTTACCG       8.0\n",
      "Name: n_total_fragments, Length: 73652, dtype: float64\n",
      "AAACTACCAGAAACCCGAGATA    0.250000\n",
      "AAACTACCAGAAACCTAAGTGG    0.178571\n",
      "AAACTACCAGAAACGGATCAGT    0.263158\n",
      "AAACTACCAGAAACGTCCCGTT    0.246201\n",
      "AAACTACCAGAAACTAGCCCTA    0.315789\n",
      "                            ...   \n",
      "TTCCATCCTCTTTCGCGTGTAA    0.166667\n",
      "TTCCATCCTCTTTCTGCAGACT    0.428571\n",
      "TTCCATCCTCTTTCTGGCGCAG    0.444444\n",
      "TTCCATCCTCTTTGCCGGAAGG    0.200000\n",
      "TTCCATCCTCTTTGTGTTACCG    0.000000\n",
      "Name: pct_fragments_in_promoters, Length: 73652, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "for entry in adata.obs:\n",
    "    print(adata.obs[entry])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "33800239",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "Can't implicitly convert non-string objects to strings\n\nAbove error raised while writing key 'Mean-Fragment-Length' of <class 'h5py._hl.group.Group'> to /",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mTypeError\u001B[0m                                 Traceback (most recent call last)",
      "File \u001B[0;32m~/.conda/envs/sctoolbox/lib/python3.10/site-packages/anndata/_io/utils.py:214\u001B[0m, in \u001B[0;36mreport_write_key_on_error.<locals>.func_wrapper\u001B[0;34m(elem, key, val, *args, **kwargs)\u001B[0m\n\u001B[1;32m    213\u001B[0m \u001B[38;5;28;01mtry\u001B[39;00m:\n\u001B[0;32m--> 214\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[43mfunc\u001B[49m\u001B[43m(\u001B[49m\u001B[43melem\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mkey\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mval\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[43margs\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[43mkwargs\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m    215\u001B[0m \u001B[38;5;28;01mexcept\u001B[39;00m \u001B[38;5;167;01mException\u001B[39;00m \u001B[38;5;28;01mas\u001B[39;00m e:\n",
      "File \u001B[0;32m~/.conda/envs/sctoolbox/lib/python3.10/site-packages/anndata/_io/specs/registry.py:171\u001B[0m, in \u001B[0;36mwrite_elem\u001B[0;34m(f, k, elem, modifiers, *args, **kwargs)\u001B[0m\n\u001B[1;32m    167\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m (\n\u001B[1;32m    168\u001B[0m     \u001B[38;5;28mhasattr\u001B[39m(elem, \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mdtype\u001B[39m\u001B[38;5;124m\"\u001B[39m)\n\u001B[1;32m    169\u001B[0m     \u001B[38;5;129;01mand\u001B[39;00m (dest_type, (t, elem\u001B[38;5;241m.\u001B[39mdtype\u001B[38;5;241m.\u001B[39mkind), modifiers) \u001B[38;5;129;01min\u001B[39;00m _REGISTRY\u001B[38;5;241m.\u001B[39mwrite\n\u001B[1;32m    170\u001B[0m ):\n\u001B[0;32m--> 171\u001B[0m     \u001B[43m_REGISTRY\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mget_writer\u001B[49m\u001B[43m(\u001B[49m\u001B[43mdest_type\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43m(\u001B[49m\u001B[43mt\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43melem\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mdtype\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mkind\u001B[49m\u001B[43m)\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mmodifiers\u001B[49m\u001B[43m)\u001B[49m\u001B[43m(\u001B[49m\n\u001B[1;32m    172\u001B[0m \u001B[43m        \u001B[49m\u001B[43mf\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mk\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43melem\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[43margs\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[43mkwargs\u001B[49m\n\u001B[1;32m    173\u001B[0m \u001B[43m    \u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m    174\u001B[0m \u001B[38;5;28;01melse\u001B[39;00m:\n",
      "File \u001B[0;32m~/.conda/envs/sctoolbox/lib/python3.10/site-packages/anndata/_io/specs/registry.py:24\u001B[0m, in \u001B[0;36mwrite_spec.<locals>.decorator.<locals>.wrapper\u001B[0;34m(g, k, *args, **kwargs)\u001B[0m\n\u001B[1;32m     22\u001B[0m \u001B[38;5;129m@wraps\u001B[39m(func)\n\u001B[1;32m     23\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m \u001B[38;5;21mwrapper\u001B[39m(g, k, \u001B[38;5;241m*\u001B[39margs, \u001B[38;5;241m*\u001B[39m\u001B[38;5;241m*\u001B[39mkwargs):\n\u001B[0;32m---> 24\u001B[0m     result \u001B[38;5;241m=\u001B[39m \u001B[43mfunc\u001B[49m\u001B[43m(\u001B[49m\u001B[43mg\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mk\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[43margs\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[43mkwargs\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m     25\u001B[0m     g[k]\u001B[38;5;241m.\u001B[39mattrs\u001B[38;5;241m.\u001B[39msetdefault(\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mencoding-type\u001B[39m\u001B[38;5;124m\"\u001B[39m, spec\u001B[38;5;241m.\u001B[39mencoding_type)\n",
      "File \u001B[0;32m~/.conda/envs/sctoolbox/lib/python3.10/site-packages/anndata/_io/specs/methods.py:346\u001B[0m, in \u001B[0;36mwrite_vlen_string_array\u001B[0;34m(f, k, elem, dataset_kwargs)\u001B[0m\n\u001B[1;32m    345\u001B[0m str_dtype \u001B[38;5;241m=\u001B[39m h5py\u001B[38;5;241m.\u001B[39mspecial_dtype(vlen\u001B[38;5;241m=\u001B[39m\u001B[38;5;28mstr\u001B[39m)\n\u001B[0;32m--> 346\u001B[0m \u001B[43mf\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mcreate_dataset\u001B[49m\u001B[43m(\u001B[49m\u001B[43mk\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mdata\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43melem\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mastype\u001B[49m\u001B[43m(\u001B[49m\u001B[43mstr_dtype\u001B[49m\u001B[43m)\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mdtype\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mstr_dtype\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[43mdataset_kwargs\u001B[49m\u001B[43m)\u001B[49m\n",
      "File \u001B[0;32m~/.conda/envs/sctoolbox/lib/python3.10/site-packages/h5py/_hl/group.py:183\u001B[0m, in \u001B[0;36mGroup.create_dataset\u001B[0;34m(self, name, shape, dtype, data, **kwds)\u001B[0m\n\u001B[1;32m    181\u001B[0m         group \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mrequire_group(parent_path)\n\u001B[0;32m--> 183\u001B[0m dsid \u001B[38;5;241m=\u001B[39m \u001B[43mdataset\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mmake_new_dset\u001B[49m\u001B[43m(\u001B[49m\u001B[43mgroup\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mshape\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mdtype\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mdata\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mname\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[43mkwds\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m    184\u001B[0m dset \u001B[38;5;241m=\u001B[39m dataset\u001B[38;5;241m.\u001B[39mDataset(dsid)\n",
      "File \u001B[0;32m~/.conda/envs/sctoolbox/lib/python3.10/site-packages/h5py/_hl/dataset.py:168\u001B[0m, in \u001B[0;36mmake_new_dset\u001B[0;34m(parent, shape, dtype, data, name, chunks, compression, shuffle, fletcher32, maxshape, compression_opts, fillvalue, scaleoffset, track_times, external, track_order, dcpl, dapl, efile_prefix, virtual_prefix, allow_unknown_filter, rdcc_nslots, rdcc_nbytes, rdcc_w0)\u001B[0m\n\u001B[1;32m    167\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m (data \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m) \u001B[38;5;129;01mand\u001B[39;00m (\u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28misinstance\u001B[39m(data, Empty)):\n\u001B[0;32m--> 168\u001B[0m     \u001B[43mdset_id\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mwrite\u001B[49m\u001B[43m(\u001B[49m\u001B[43mh5s\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mALL\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mh5s\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mALL\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mdata\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m    170\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m dset_id\n",
      "File \u001B[0;32mh5py/_objects.pyx:54\u001B[0m, in \u001B[0;36mh5py._objects.with_phil.wrapper\u001B[0;34m()\u001B[0m\n",
      "File \u001B[0;32mh5py/_objects.pyx:55\u001B[0m, in \u001B[0;36mh5py._objects.with_phil.wrapper\u001B[0;34m()\u001B[0m\n",
      "File \u001B[0;32mh5py/h5d.pyx:280\u001B[0m, in \u001B[0;36mh5py.h5d.DatasetID.write\u001B[0;34m()\u001B[0m\n",
      "File \u001B[0;32mh5py/_proxy.pyx:145\u001B[0m, in \u001B[0;36mh5py._proxy.dset_rw\u001B[0;34m()\u001B[0m\n",
      "File \u001B[0;32mh5py/_conv.pyx:444\u001B[0m, in \u001B[0;36mh5py._conv.str2vlen\u001B[0;34m()\u001B[0m\n",
      "File \u001B[0;32mh5py/_conv.pyx:95\u001B[0m, in \u001B[0;36mh5py._conv.generic_converter\u001B[0;34m()\u001B[0m\n",
      "File \u001B[0;32mh5py/_conv.pyx:249\u001B[0m, in \u001B[0;36mh5py._conv.conv_str2vlen\u001B[0;34m()\u001B[0m\n",
      "\u001B[0;31mTypeError\u001B[0m: Can't implicitly convert non-string objects to strings",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001B[0;31mTypeError\u001B[0m                                 Traceback (most recent call last)",
      "Cell \u001B[0;32mIn[45], line 6\u001B[0m\n\u001B[1;32m      3\u001B[0m cr\u001B[38;5;241m.\u001B[39mbuild_infor(adata, \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mAnndata_path\u001B[39m\u001B[38;5;124m\"\u001B[39m, output_dir)\n\u001B[1;32m      5\u001B[0m adata_output \u001B[38;5;241m=\u001B[39m tree\u001B[38;5;241m.\u001B[39massembled_anndata\n\u001B[0;32m----> 6\u001B[0m \u001B[43madata\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mwrite\u001B[49m\u001B[43m(\u001B[49m\u001B[43mfilename\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43madata_output\u001B[49m\u001B[43m)\u001B[49m\n",
      "File \u001B[0;32m~/.conda/envs/sctoolbox/lib/python3.10/site-packages/anndata/_core/anndata.py:1918\u001B[0m, in \u001B[0;36mAnnData.write_h5ad\u001B[0;34m(self, filename, compression, compression_opts, force_dense, as_dense)\u001B[0m\n\u001B[1;32m   1915\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m filename \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m:\n\u001B[1;32m   1916\u001B[0m     filename \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mfilename\n\u001B[0;32m-> 1918\u001B[0m \u001B[43m_write_h5ad\u001B[49m\u001B[43m(\u001B[49m\n\u001B[1;32m   1919\u001B[0m \u001B[43m    \u001B[49m\u001B[43mPath\u001B[49m\u001B[43m(\u001B[49m\u001B[43mfilename\u001B[49m\u001B[43m)\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m   1920\u001B[0m \u001B[43m    \u001B[49m\u001B[38;5;28;43mself\u001B[39;49m\u001B[43m,\u001B[49m\n\u001B[1;32m   1921\u001B[0m \u001B[43m    \u001B[49m\u001B[43mcompression\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mcompression\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m   1922\u001B[0m \u001B[43m    \u001B[49m\u001B[43mcompression_opts\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mcompression_opts\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m   1923\u001B[0m \u001B[43m    \u001B[49m\u001B[43mforce_dense\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mforce_dense\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m   1924\u001B[0m \u001B[43m    \u001B[49m\u001B[43mas_dense\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mas_dense\u001B[49m\u001B[43m,\u001B[49m\n\u001B[1;32m   1925\u001B[0m \u001B[43m\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m   1927\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39misbacked:\n\u001B[1;32m   1928\u001B[0m     \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mfile\u001B[38;5;241m.\u001B[39mfilename \u001B[38;5;241m=\u001B[39m filename\n",
      "File \u001B[0;32m~/.conda/envs/sctoolbox/lib/python3.10/site-packages/anndata/_io/h5ad.py:98\u001B[0m, in \u001B[0;36mwrite_h5ad\u001B[0;34m(filepath, adata, force_dense, as_dense, dataset_kwargs, **kwargs)\u001B[0m\n\u001B[1;32m     96\u001B[0m \u001B[38;5;28;01melif\u001B[39;00m adata\u001B[38;5;241m.\u001B[39mraw \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m:\n\u001B[1;32m     97\u001B[0m     write_elem(f, \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mraw\u001B[39m\u001B[38;5;124m\"\u001B[39m, adata\u001B[38;5;241m.\u001B[39mraw, dataset_kwargs\u001B[38;5;241m=\u001B[39mdataset_kwargs)\n\u001B[0;32m---> 98\u001B[0m \u001B[43mwrite_elem\u001B[49m\u001B[43m(\u001B[49m\u001B[43mf\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[38;5;124;43mobs\u001B[39;49m\u001B[38;5;124;43m\"\u001B[39;49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43madata\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mobs\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mdataset_kwargs\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mdataset_kwargs\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m     99\u001B[0m write_elem(f, \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mvar\u001B[39m\u001B[38;5;124m\"\u001B[39m, adata\u001B[38;5;241m.\u001B[39mvar, dataset_kwargs\u001B[38;5;241m=\u001B[39mdataset_kwargs)\n\u001B[1;32m    100\u001B[0m write_elem(f, \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mobsm\u001B[39m\u001B[38;5;124m\"\u001B[39m, \u001B[38;5;28mdict\u001B[39m(adata\u001B[38;5;241m.\u001B[39mobsm), dataset_kwargs\u001B[38;5;241m=\u001B[39mdataset_kwargs)\n",
      "File \u001B[0;32m~/.conda/envs/sctoolbox/lib/python3.10/site-packages/anndata/_io/utils.py:214\u001B[0m, in \u001B[0;36mreport_write_key_on_error.<locals>.func_wrapper\u001B[0;34m(elem, key, val, *args, **kwargs)\u001B[0m\n\u001B[1;32m    211\u001B[0m \u001B[38;5;129m@wraps\u001B[39m(func)\n\u001B[1;32m    212\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m \u001B[38;5;21mfunc_wrapper\u001B[39m(elem, key, val, \u001B[38;5;241m*\u001B[39margs, \u001B[38;5;241m*\u001B[39m\u001B[38;5;241m*\u001B[39mkwargs):\n\u001B[1;32m    213\u001B[0m     \u001B[38;5;28;01mtry\u001B[39;00m:\n\u001B[0;32m--> 214\u001B[0m         \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[43mfunc\u001B[49m\u001B[43m(\u001B[49m\u001B[43melem\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mkey\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mval\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[43margs\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[43mkwargs\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m    215\u001B[0m     \u001B[38;5;28;01mexcept\u001B[39;00m \u001B[38;5;167;01mException\u001B[39;00m \u001B[38;5;28;01mas\u001B[39;00m e:\n\u001B[1;32m    216\u001B[0m         \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mAbove error raised while writing key\u001B[39m\u001B[38;5;124m\"\u001B[39m \u001B[38;5;129;01min\u001B[39;00m \u001B[38;5;28mformat\u001B[39m(e):\n",
      "File \u001B[0;32m~/.conda/envs/sctoolbox/lib/python3.10/site-packages/anndata/_io/specs/registry.py:175\u001B[0m, in \u001B[0;36mwrite_elem\u001B[0;34m(f, k, elem, modifiers, *args, **kwargs)\u001B[0m\n\u001B[1;32m    171\u001B[0m     _REGISTRY\u001B[38;5;241m.\u001B[39mget_writer(dest_type, (t, elem\u001B[38;5;241m.\u001B[39mdtype\u001B[38;5;241m.\u001B[39mkind), modifiers)(\n\u001B[1;32m    172\u001B[0m         f, k, elem, \u001B[38;5;241m*\u001B[39margs, \u001B[38;5;241m*\u001B[39m\u001B[38;5;241m*\u001B[39mkwargs\n\u001B[1;32m    173\u001B[0m     )\n\u001B[1;32m    174\u001B[0m \u001B[38;5;28;01melse\u001B[39;00m:\n\u001B[0;32m--> 175\u001B[0m     \u001B[43m_REGISTRY\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mget_writer\u001B[49m\u001B[43m(\u001B[49m\u001B[43mdest_type\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mt\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mmodifiers\u001B[49m\u001B[43m)\u001B[49m\u001B[43m(\u001B[49m\u001B[43mf\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mk\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43melem\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[43margs\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[43mkwargs\u001B[49m\u001B[43m)\u001B[49m\n",
      "File \u001B[0;32m~/.conda/envs/sctoolbox/lib/python3.10/site-packages/anndata/_io/specs/registry.py:24\u001B[0m, in \u001B[0;36mwrite_spec.<locals>.decorator.<locals>.wrapper\u001B[0;34m(g, k, *args, **kwargs)\u001B[0m\n\u001B[1;32m     22\u001B[0m \u001B[38;5;129m@wraps\u001B[39m(func)\n\u001B[1;32m     23\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m \u001B[38;5;21mwrapper\u001B[39m(g, k, \u001B[38;5;241m*\u001B[39margs, \u001B[38;5;241m*\u001B[39m\u001B[38;5;241m*\u001B[39mkwargs):\n\u001B[0;32m---> 24\u001B[0m     result \u001B[38;5;241m=\u001B[39m \u001B[43mfunc\u001B[49m\u001B[43m(\u001B[49m\u001B[43mg\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mk\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[43margs\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[43mkwargs\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m     25\u001B[0m     g[k]\u001B[38;5;241m.\u001B[39mattrs\u001B[38;5;241m.\u001B[39msetdefault(\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mencoding-type\u001B[39m\u001B[38;5;124m\"\u001B[39m, spec\u001B[38;5;241m.\u001B[39mencoding_type)\n\u001B[1;32m     26\u001B[0m     g[k]\u001B[38;5;241m.\u001B[39mattrs\u001B[38;5;241m.\u001B[39msetdefault(\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mencoding-version\u001B[39m\u001B[38;5;124m\"\u001B[39m, spec\u001B[38;5;241m.\u001B[39mencoding_version)\n",
      "File \u001B[0;32m~/.conda/envs/sctoolbox/lib/python3.10/site-packages/anndata/_io/specs/methods.py:514\u001B[0m, in \u001B[0;36mwrite_dataframe\u001B[0;34m(f, key, df, dataset_kwargs)\u001B[0m\n\u001B[1;32m    511\u001B[0m write_elem(group, index_name, df\u001B[38;5;241m.\u001B[39mindex\u001B[38;5;241m.\u001B[39m_values, dataset_kwargs\u001B[38;5;241m=\u001B[39mdataset_kwargs)\n\u001B[1;32m    512\u001B[0m \u001B[38;5;28;01mfor\u001B[39;00m colname, series \u001B[38;5;129;01min\u001B[39;00m df\u001B[38;5;241m.\u001B[39mitems():\n\u001B[1;32m    513\u001B[0m     \u001B[38;5;66;03m# TODO: this should write the \"true\" representation of the series (i.e. the underlying array or ndarray depending)\u001B[39;00m\n\u001B[0;32m--> 514\u001B[0m     \u001B[43mwrite_elem\u001B[49m\u001B[43m(\u001B[49m\u001B[43mgroup\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mcolname\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mseries\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43m_values\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mdataset_kwargs\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[43mdataset_kwargs\u001B[49m\u001B[43m)\u001B[49m\n",
      "File \u001B[0;32m~/.conda/envs/sctoolbox/lib/python3.10/site-packages/anndata/_io/utils.py:220\u001B[0m, in \u001B[0;36mreport_write_key_on_error.<locals>.func_wrapper\u001B[0;34m(elem, key, val, *args, **kwargs)\u001B[0m\n\u001B[1;32m    218\u001B[0m \u001B[38;5;28;01melse\u001B[39;00m:\n\u001B[1;32m    219\u001B[0m     parent \u001B[38;5;241m=\u001B[39m _get_parent(elem)\n\u001B[0;32m--> 220\u001B[0m     \u001B[38;5;28;01mraise\u001B[39;00m \u001B[38;5;28mtype\u001B[39m(e)(\n\u001B[1;32m    221\u001B[0m         \u001B[38;5;124mf\u001B[39m\u001B[38;5;124m\"\u001B[39m\u001B[38;5;132;01m{\u001B[39;00me\u001B[38;5;132;01m}\u001B[39;00m\u001B[38;5;130;01m\\n\u001B[39;00m\u001B[38;5;130;01m\\n\u001B[39;00m\u001B[38;5;124m\"\u001B[39m\n\u001B[1;32m    222\u001B[0m         \u001B[38;5;124mf\u001B[39m\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mAbove error raised while writing key \u001B[39m\u001B[38;5;132;01m{\u001B[39;00mkey\u001B[38;5;132;01m!r}\u001B[39;00m\u001B[38;5;124m of \u001B[39m\u001B[38;5;132;01m{\u001B[39;00m\u001B[38;5;28mtype\u001B[39m(elem)\u001B[38;5;132;01m}\u001B[39;00m\u001B[38;5;124m \u001B[39m\u001B[38;5;124m\"\u001B[39m\n\u001B[1;32m    223\u001B[0m         \u001B[38;5;124mf\u001B[39m\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mto \u001B[39m\u001B[38;5;132;01m{\u001B[39;00mparent\u001B[38;5;132;01m}\u001B[39;00m\u001B[38;5;124m\"\u001B[39m\n\u001B[1;32m    224\u001B[0m     ) \u001B[38;5;28;01mfrom\u001B[39;00m \u001B[38;5;21;01me\u001B[39;00m\n",
      "\u001B[0;31mTypeError\u001B[0m: Can't implicitly convert non-string objects to strings\n\nAbove error raised while writing key 'Mean-Fragment-Length' of <class 'h5py._hl.group.Group'> to /"
     ]
    }
   ],
   "source": [
    "#Saving the data\n",
    "cr.build_infor(adata, \"Test_number\", test)\n",
    "cr.build_infor(adata, \"Anndata_path\", output_dir)\n",
    "\n",
    "adata_output = tree.assembled_anndata\n",
    "adata.write(filename=adata_output)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9437c45",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "repo_path = os.getcwd()\n",
    "notebook_name = '1_assembling.ipynb'\n",
    "notebook_path = os.path.join(repo_path, notebook_name)\n",
    "notebook_copy = os.path.join(tree.assemble_dir , notebook_name)\n",
    "shutil.copyfile(notebook_path, notebook_copy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "779d98ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AAACTACCAGAAACCCGAGATA    [0]\n",
       "AAACTACCAGAAACCTAAGTGG    [0]\n",
       "AAACTACCAGAAACGGATCAGT    [0]\n",
       "AAACTACCAGAAACGTCCCGTT    [0]\n",
       "AAACTACCAGAAACTAGCCCTA    [0]\n",
       "                         ... \n",
       "TTCCATCCTCTTTCGCGTGTAA    [0]\n",
       "TTCCATCCTCTTTCTGCAGACT    [0]\n",
       "TTCCATCCTCTTTCTGGCGCAG    [0]\n",
       "TTCCATCCTCTTTGCCGGAAGG     []\n",
       "TTCCATCCTCTTTGTGTTACCG    [0]\n",
       "Name: Maxima, Length: 73652, dtype: category\n",
       "Categories (89, object): ['[0, 3]', '[0, 4]', '[0, 5]', '[0, 6, 9]', ..., '[6]', '[9]', '[]', 'nan']"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.pop(\"Maxima\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "3c7a489b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TN</th>\n",
       "      <th>UM</th>\n",
       "      <th>PP</th>\n",
       "      <th>UQ</th>\n",
       "      <th>CM</th>\n",
       "      <th>file</th>\n",
       "      <th>sample</th>\n",
       "      <th>Sample_new</th>\n",
       "      <th>n_features_by_counts</th>\n",
       "      <th>log1p_n_features_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>Mean-Fragment-Length</th>\n",
       "      <th>Median-Fragment-Length</th>\n",
       "      <th>Fragment-Count</th>\n",
       "      <th>Maxima-Count</th>\n",
       "      <th>nucleosomal-score</th>\n",
       "      <th>n_fragments_in_promoters</th>\n",
       "      <th>n_total_fragments</th>\n",
       "      <th>pct_fragments_in_promoters</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACCCGAGATA</th>\n",
       "      <td>33.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>5</td>\n",
       "      <td>1.791759</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.791759</td>\n",
       "      <td>82.36</td>\n",
       "      <td>49</td>\n",
       "      <td>11.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>145.544305</td>\n",
       "      <td>4.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACCTAAGTGG</th>\n",
       "      <td>52.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>17</td>\n",
       "      <td>2.890372</td>\n",
       "      <td>17.0</td>\n",
       "      <td>2.890372</td>\n",
       "      <td>170.38</td>\n",
       "      <td>138.5</td>\n",
       "      <td>26.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>102.513844</td>\n",
       "      <td>5.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0.178571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACGGATCAGT</th>\n",
       "      <td>27.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>10</td>\n",
       "      <td>2.397895</td>\n",
       "      <td>10.0</td>\n",
       "      <td>2.397895</td>\n",
       "      <td>161.17</td>\n",
       "      <td>74.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>140.045502</td>\n",
       "      <td>5.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.263158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACGTCCCGTT</th>\n",
       "      <td>6223.0</td>\n",
       "      <td>5231.0</td>\n",
       "      <td>5213.0</td>\n",
       "      <td>3779.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>1648</td>\n",
       "      <td>7.407924</td>\n",
       "      <td>1791.0</td>\n",
       "      <td>7.491087</td>\n",
       "      <td>144.03</td>\n",
       "      <td>106</td>\n",
       "      <td>3475.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>19.623836</td>\n",
       "      <td>1296.0</td>\n",
       "      <td>5264.0</td>\n",
       "      <td>0.246201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACTACCAGAAACTAGCCCTA</th>\n",
       "      <td>41.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>13</td>\n",
       "      <td>2.639057</td>\n",
       "      <td>13.0</td>\n",
       "      <td>2.639057</td>\n",
       "      <td>160.5</td>\n",
       "      <td>131.5</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>152.003069</td>\n",
       "      <td>6.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.315789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTCGCGTGTAA</th>\n",
       "      <td>9.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>3</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>125.8</td>\n",
       "      <td>94</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>220.573902</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTCTGCAGACT</th>\n",
       "      <td>10.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>3</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>87.0</td>\n",
       "      <td>58</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>324.955404</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.428571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTCTGGCGCAG</th>\n",
       "      <td>17.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>86.0</td>\n",
       "      <td>46</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>324.955404</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.444444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTGCCGGAAGG</th>\n",
       "      <td>8.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>5</td>\n",
       "      <td>1.791759</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.791759</td>\n",
       "      <td>163.5</td>\n",
       "      <td>163.5</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCCATCCTCTTTGTGTTACCG</th>\n",
       "      <td>19.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>/mnt/workspace/jdetlef/data/anndata/esophagus_...</td>\n",
       "      <td>MB</td>\n",
       "      <td>sample1</td>\n",
       "      <td>2</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>115.0</td>\n",
       "      <td>100</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>220.573902</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>73652 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            TN      UM      PP      UQ    CM  \\\n",
       "AAACTACCAGAAACCCGAGATA    33.0    18.0    18.0    15.0   0.0   \n",
       "AAACTACCAGAAACCTAAGTGG    52.0    36.0    35.0    32.0   1.0   \n",
       "AAACTACCAGAAACGGATCAGT    27.0    19.0    19.0    19.0   0.0   \n",
       "AAACTACCAGAAACGTCCCGTT  6223.0  5231.0  5213.0  3779.0  18.0   \n",
       "AAACTACCAGAAACTAGCCCTA    41.0    29.0    29.0    26.0   0.0   \n",
       "...                        ...     ...     ...     ...   ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA     9.0     6.0     6.0     6.0   0.0   \n",
       "TTCCATCCTCTTTCTGCAGACT    10.0     6.0     6.0     6.0   1.0   \n",
       "TTCCATCCTCTTTCTGGCGCAG    17.0     5.0     5.0     5.0   1.0   \n",
       "TTCCATCCTCTTTGCCGGAAGG     8.0     7.0     7.0     7.0   0.0   \n",
       "TTCCATCCTCTTTGTGTTACCG    19.0     9.0     9.0     7.0   0.0   \n",
       "\n",
       "                                                                     file  \\\n",
       "AAACTACCAGAAACCCGAGATA  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "AAACTACCAGAAACCTAAGTGG  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "AAACTACCAGAAACGGATCAGT  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "AAACTACCAGAAACGTCCCGTT  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "AAACTACCAGAAACTAGCCCTA  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "...                                                                   ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "TTCCATCCTCTTTCTGCAGACT  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "TTCCATCCTCTTTCTGGCGCAG  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "TTCCATCCTCTTTGCCGGAAGG  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "TTCCATCCTCTTTGTGTTACCG  /mnt/workspace/jdetlef/data/anndata/esophagus_...   \n",
       "\n",
       "                       sample Sample_new  n_features_by_counts  \\\n",
       "AAACTACCAGAAACCCGAGATA     MB    sample1                     5   \n",
       "AAACTACCAGAAACCTAAGTGG     MB    sample1                    17   \n",
       "AAACTACCAGAAACGGATCAGT     MB    sample1                    10   \n",
       "AAACTACCAGAAACGTCCCGTT     MB    sample1                  1648   \n",
       "AAACTACCAGAAACTAGCCCTA     MB    sample1                    13   \n",
       "...                       ...        ...                   ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA     MB    sample1                     3   \n",
       "TTCCATCCTCTTTCTGCAGACT     MB    sample1                     3   \n",
       "TTCCATCCTCTTTCTGGCGCAG     MB    sample1                     0   \n",
       "TTCCATCCTCTTTGCCGGAAGG     MB    sample1                     5   \n",
       "TTCCATCCTCTTTGTGTTACCG     MB    sample1                     2   \n",
       "\n",
       "                        log1p_n_features_by_counts  total_counts  \\\n",
       "AAACTACCAGAAACCCGAGATA                    1.791759           5.0   \n",
       "AAACTACCAGAAACCTAAGTGG                    2.890372          17.0   \n",
       "AAACTACCAGAAACGGATCAGT                    2.397895          10.0   \n",
       "AAACTACCAGAAACGTCCCGTT                    7.407924        1791.0   \n",
       "AAACTACCAGAAACTAGCCCTA                    2.639057          13.0   \n",
       "...                                            ...           ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA                    1.386294           3.0   \n",
       "TTCCATCCTCTTTCTGCAGACT                    1.386294           3.0   \n",
       "TTCCATCCTCTTTCTGGCGCAG                    0.000000           0.0   \n",
       "TTCCATCCTCTTTGCCGGAAGG                    1.791759           5.0   \n",
       "TTCCATCCTCTTTGTGTTACCG                    1.098612           2.0   \n",
       "\n",
       "                        log1p_total_counts Mean-Fragment-Length  \\\n",
       "AAACTACCAGAAACCCGAGATA            1.791759                82.36   \n",
       "AAACTACCAGAAACCTAAGTGG            2.890372               170.38   \n",
       "AAACTACCAGAAACGGATCAGT            2.397895               161.17   \n",
       "AAACTACCAGAAACGTCCCGTT            7.491087               144.03   \n",
       "AAACTACCAGAAACTAGCCCTA            2.639057                160.5   \n",
       "...                                    ...                  ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA            1.386294                125.8   \n",
       "TTCCATCCTCTTTCTGCAGACT            1.386294                 87.0   \n",
       "TTCCATCCTCTTTCTGGCGCAG            0.000000                 86.0   \n",
       "TTCCATCCTCTTTGCCGGAAGG            1.791759                163.5   \n",
       "TTCCATCCTCTTTGTGTTACCG            1.098612                115.0   \n",
       "\n",
       "                       Median-Fragment-Length  Fragment-Count  Maxima-Count  \\\n",
       "AAACTACCAGAAACCCGAGATA                     49            11.0           1.0   \n",
       "AAACTACCAGAAACCTAAGTGG                  138.5            26.0           1.0   \n",
       "AAACTACCAGAAACGGATCAGT                   74.0            12.0           1.0   \n",
       "AAACTACCAGAAACGTCCCGTT                    106          3475.0           1.0   \n",
       "AAACTACCAGAAACTAGCCCTA                  131.5            10.0           1.0   \n",
       "...                                       ...             ...           ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA                     94             5.0           1.0   \n",
       "TTCCATCCTCTTTCTGCAGACT                     58             3.0           1.0   \n",
       "TTCCATCCTCTTTCTGGCGCAG                     46             3.0           1.0   \n",
       "TTCCATCCTCTTTGCCGGAAGG                  163.5             2.0           0.0   \n",
       "TTCCATCCTCTTTGTGTTACCG                    100             5.0           1.0   \n",
       "\n",
       "                        nucleosomal-score  n_fragments_in_promoters  \\\n",
       "AAACTACCAGAAACCCGAGATA         145.544305                       4.0   \n",
       "AAACTACCAGAAACCTAAGTGG         102.513844                       5.0   \n",
       "AAACTACCAGAAACGGATCAGT         140.045502                       5.0   \n",
       "AAACTACCAGAAACGTCCCGTT          19.623836                    1296.0   \n",
       "AAACTACCAGAAACTAGCCCTA         152.003069                       6.0   \n",
       "...                                   ...                       ...   \n",
       "TTCCATCCTCTTTCGCGTGTAA         220.573902                       1.0   \n",
       "TTCCATCCTCTTTCTGCAGACT         324.955404                       3.0   \n",
       "TTCCATCCTCTTTCTGGCGCAG         324.955404                       4.0   \n",
       "TTCCATCCTCTTTGCCGGAAGG                inf                       1.0   \n",
       "TTCCATCCTCTTTGTGTTACCG         220.573902                       0.0   \n",
       "\n",
       "                        n_total_fragments  pct_fragments_in_promoters  \n",
       "AAACTACCAGAAACCCGAGATA               16.0                    0.250000  \n",
       "AAACTACCAGAAACCTAAGTGG               28.0                    0.178571  \n",
       "AAACTACCAGAAACGGATCAGT               19.0                    0.263158  \n",
       "AAACTACCAGAAACGTCCCGTT             5264.0                    0.246201  \n",
       "AAACTACCAGAAACTAGCCCTA               19.0                    0.315789  \n",
       "...                                   ...                         ...  \n",
       "TTCCATCCTCTTTCGCGTGTAA                6.0                    0.166667  \n",
       "TTCCATCCTCTTTCTGCAGACT                7.0                    0.428571  \n",
       "TTCCATCCTCTTTCTGGCGCAG                9.0                    0.444444  \n",
       "TTCCATCCTCTTTGCCGGAAGG                5.0                    0.200000  \n",
       "TTCCATCCTCTTTGTGTTACCG                8.0                    0.000000  \n",
       "\n",
       "[73652 rows x 20 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs >= "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea0eec2c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sctoolbox",
   "language": "python",
   "name": "sctoolbox"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
